{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Manduca sexta Proteomics\n",
    "\n",
    "Developmental stages of hemolymph  \n",
    "cell membrane of fat body and hemocytes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate Protein seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connect MsAnnotated sequences to OGS2 and MCOT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "# annotated sequence name to OGS2 and MCOT\n",
    "file_Anno = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\raw\\20170627AnnotatedManduca.fa\"\n",
    "file_mcotseq = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta\"\n",
    "file_ogs2seq = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\ms_ogs_proteins.fa\"\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import sys\n",
    "sys.path.append(r'C:\\Users\\ATPs\\Documents\\GitHub\\XCProject\\MCuNovo\\SeqCompare')\n",
    "import SeqCompare\n",
    "\n",
    "df_Anno2OGS = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_ogs2seq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2MCOT = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_mcotseq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2OGS['seq1_id'] = df_Anno2OGS['seq1_id'].apply(lambda x:x.replace('MsKnown',''))\n",
    "df_Anno2MCOT['seq1_id'] = df_Anno2MCOT['seq1_id'].apply(lambda x:x.replace('MsKnown',''))\n",
    "df_Anno2OGS['MS'] = df_Anno2OGS.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "df_Anno2MCOT['MS'] = df_Anno2MCOT.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "\n",
    "excel = pd.ExcelWriter(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsAnnotatedSeq.xlsx\", engine='xlsxwriter')\n",
    "df_Anno2OGS = df_Anno2OGS.sort_values(by=['seq1_id','MS'],ascending=False)\n",
    "df_Anno2OGS1 = df_Anno2OGS.drop_duplicates(subset=['seq1_id'],keep='first')\n",
    "df_Anno2OGS1 = df_Anno2OGS1[df_Anno2OGS1['MS']>0.4]\n",
    "df_Anno2OGS1 = df_Anno2OGS1.sort_index()\n",
    "df_Anno2OGS1 = df_Anno2OGS1.set_index('seq1_id')\n",
    "df_Anno2OGS1.to_excel(excel,sheet_name='Anno2OGS')\n",
    "df_Anno2OGS = df_Anno2OGS.sort_values(by=['seq2_id','MS'],ascending=False)\n",
    "df_Anno2OGS1 = df_Anno2OGS.drop_duplicates(subset=['seq2_id'],keep='first')\n",
    "df_Anno2OGS1 = df_Anno2OGS1[df_Anno2OGS1['MS']>0.4]\n",
    "df_Anno2OGS1 = df_Anno2OGS1.sort_index()\n",
    "df_Anno2OGS1 = df_Anno2OGS1.set_index('seq2_id')\n",
    "df_Anno2OGS1.to_excel(excel,sheet_name='OGS2Anno')\n",
    "df_Anno2MCOT = df_Anno2MCOT.sort_values(by=['seq1_id','MS'],ascending=False)\n",
    "df_Anno2MCOT1 = df_Anno2MCOT.drop_duplicates(subset=['seq1_id'],keep='first')\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1[df_Anno2MCOT1['MS']>0.4]\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.sort_index()\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.set_index('seq1_id')\n",
    "df_Anno2MCOT1.to_excel(excel,sheet_name='Anno2MCOT')\n",
    "df_Anno2MCOT = df_Anno2MCOT.sort_values(by=['seq2_id','MS'],ascending=False)\n",
    "df_Anno2MCOT1 = df_Anno2MCOT.drop_duplicates(subset=['seq2_id'],keep='first')\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1[df_Anno2MCOT1['MS']>0.4]\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.sort_index()\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.set_index('seq2_id')\n",
    "df_Anno2MCOT1.to_excel(excel,sheet_name='MCOT2Anno')\n",
    "excel.save()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "results were further improved manually. with OGS2 and MCOT as the reference, only keep ones with highest MS with annotated seqeunces."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Jiang want to add additional SPSPHs to the known ones\n",
    "20191130  update the known sequences and the excel file.\n",
    "\n",
    "```python\n",
    "# annotated sequence name to OGS2 and MCOT\n",
    "file_Anno = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20170627AnnotatedManduca.fa\"\n",
    "file_mcotseq = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta\"\n",
    "file_ogs2seq = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\ms_ogs_proteins.fa\"\n",
    "file_SPSPHs = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20191130SPSPHs_JiangAdded.txt\"\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import sys\n",
    "sys.path.append(r'C:\\Users\\ATPs\\Documents\\GitHub\\XCProject\\MCuNovo\\SeqCompare')\n",
    "import SeqCompare\n",
    "\n",
    "seqs_Anno = SeqIO.to_dict(SeqIO.parse(file_Anno,'fasta'))\n",
    "seqs_SPSPH = SeqIO.to_dict(SeqIO.parse(file_SPSPHs, 'fasta'))\n",
    "\n",
    "seqs_new = [e for e in seqs_SPSPH if e not in seqs_Anno] # totally 53 sequences, and all SP(H)2xx\n",
    "ls_seqs = list(seqs_Anno.values()) + [seqs_SPSPH[e] for e in seqs_new]\n",
    "# update the file_Anno\n",
    "open(file_Anno,'w').write(''.join('>'+e.id+'\\n'+str(e.seq)+'\\n' for e in ls_seqs))\n",
    "\n",
    "\n",
    "df_Anno2OGS = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_ogs2seq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2MCOT = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_mcotseq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2OGS['seq1_id'] = df_Anno2OGS['seq1_id'].apply(lambda x:x.replace('MsKnown',''))\n",
    "df_Anno2MCOT['seq1_id'] = df_Anno2MCOT['seq1_id'].apply(lambda x:x.replace('MsKnown',''))\n",
    "df_Anno2OGS['MS'] = df_Anno2OGS.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "df_Anno2MCOT['MS'] = df_Anno2MCOT.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "\n",
    "excel = pd.ExcelWriter(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20191130MsAnnotatedSeq.xlsx\", engine='xlsxwriter')\n",
    "df_Anno2OGS = df_Anno2OGS.sort_values(by=['seq1_id','MS'],ascending=False)\n",
    "df_Anno2OGS1 = df_Anno2OGS.drop_duplicates(subset=['seq1_id'],keep='first')\n",
    "df_Anno2OGS1 = df_Anno2OGS1[df_Anno2OGS1['MS']>0.4]\n",
    "df_Anno2OGS1 = df_Anno2OGS1.sort_index()\n",
    "df_Anno2OGS1 = df_Anno2OGS1.set_index('seq1_id')\n",
    "df_Anno2OGS1.to_excel(excel,sheet_name='Anno2OGS')\n",
    "df_Anno2OGS = df_Anno2OGS.sort_values(by=['seq2_id','MS'],ascending=False)\n",
    "df_Anno2OGS1 = df_Anno2OGS.drop_duplicates(subset=['seq2_id'],keep='first')\n",
    "df_Anno2OGS1 = df_Anno2OGS1[df_Anno2OGS1['MS']>0.4]\n",
    "df_Anno2OGS1 = df_Anno2OGS1.sort_index()\n",
    "df_Anno2OGS1 = df_Anno2OGS1.set_index('seq2_id')\n",
    "df_Anno2OGS1.to_excel(excel,sheet_name='OGS2Anno')\n",
    "df_Anno2MCOT = df_Anno2MCOT.sort_values(by=['seq1_id','MS'],ascending=False)\n",
    "df_Anno2MCOT1 = df_Anno2MCOT.drop_duplicates(subset=['seq1_id'],keep='first')\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1[df_Anno2MCOT1['MS']>0.4]\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.sort_index()\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.set_index('seq1_id')\n",
    "df_Anno2MCOT1.to_excel(excel,sheet_name='Anno2MCOT')\n",
    "df_Anno2MCOT = df_Anno2MCOT.sort_values(by=['seq2_id','MS'],ascending=False)\n",
    "df_Anno2MCOT1 = df_Anno2MCOT.drop_duplicates(subset=['seq2_id'],keep='first')\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1[df_Anno2MCOT1['MS']>0.4]\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.sort_index()\n",
    "df_Anno2MCOT1 = df_Anno2MCOT1.set_index('seq2_id')\n",
    "df_Anno2MCOT1.to_excel(excel,sheet_name='MCOT2Anno')\n",
    "excel.save()\n",
    "```\n",
    "\n",
    "add the sheet `2Anno` manually. Replace the old file, rename old file as .backup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connect MsImmuneRelated sequences to OGS2 and MCOT\n",
    "\n",
    "```python\n",
    "# annotated sequence name to OGS2 and MCOT\n",
    "file_Immune = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\manduca_sexta_annotated_684_DNA_cds.fa\"\n",
    "file_mcotseq = r\"D:\\Insects\\ManducaSexta\\MCOT_OGS\\20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta\"\n",
    "file_ogs2seq = r\"D:\\Insects\\ManducaSexta\\MCOT_OGS\\ms_ogs_proteins.fa\"\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import sys\n",
    "sys.path.append(r'C:\\Users\\ATPs\\Documents\\GitHub\\XCProject\\MCuNovo\\SeqCompare')\n",
    "import SeqCompare\n",
    "\n",
    "seqs1 = list(SeqIO.parse(file_Immune, 'fasta'))\n",
    "#translate\n",
    "seqs1_aa = []\n",
    "for i in seqs1:\n",
    "    p = i.translate()\n",
    "    p.id = i.id\n",
    "    p.seq = p.seq.strip('*')\n",
    "    if len(i.seq) %3 != 0:\n",
    "        print(i.id)\n",
    "    seqs1_aa.append(p)\n",
    "#save AA\n",
    "fout = open(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\manduca_sexta_annotated_684_DNA_AA.fa\",'w')\n",
    "for s in seqs1_aa:\n",
    "    fout.write('>'+s.id+'\\n'+str(s.seq)+'\\n')\n",
    "fout.close()\n",
    "\n",
    "file_Anno = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\manduca_sexta_annotated_684_DNA_AA.fa\"\n",
    "df_Anno2OGS = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_ogs2seq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2MCOT = SeqCompare.compare2lsSeqs(seqs1=file_Anno, seqs2=file_mcotseq, identitymin = 10, outfile = None, error_rate = 0.02,muscle_exe = r\"C:\\P\\muscle\\muscle3.8.31_i86win32.exe\")\n",
    "df_Anno2OGS['MS'] = df_Anno2OGS.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "df_Anno2MCOT['MS'] = df_Anno2MCOT.apply(lambda x:x['match_len'] ** 2 /(x['seq1_len'] * x['seq2_len']),axis=1)\n",
    "df = pd.concat([df_Anno2MCOT, df_Anno2OGS], ignore_index=True)\n",
    "df = df[df['MS']>0.4]\n",
    "df.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsImmuneRelatedSeq_MCOTOGS.csv\", sep='\\t',index=None)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### annotate OGS2 and MCOT with NCBI nr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### blast against NR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "# Annotate MCOT with gene names\n",
    "#split files to 20\n",
    "from Bio import SeqIO\n",
    "filename = '/panfs/panfs.cluster/scratch/ks2073/Blast/MCOT/MCOT.fasta'\n",
    "N = 20\n",
    "fouts = [open(filename+str(i),'w') for i in range(N)]\n",
    "for n, seq in enumerate(SeqIO.parse(filename,'fasta')):\n",
    "    towrite = '>'+seq.description + '\\n'+str(seq.seq)+'\\n'\n",
    "    fouts[n % N].write(towrite)\n",
    "for i in range(N):\n",
    "    fouts[i].close()\n",
    "# generate qsub scripts\n",
    "qsubfile = '{filename}_qsub{n}'\n",
    "txt = '''#!/bin/bash\n",
    "#PBS -q batch\n",
    "#PBS -N Blastp{n}\n",
    "#PBS -l nodes=1:ppn=12\n",
    "#PBS -l walltime=80:00:00\n",
    "#PBS -m abe -M atps@outlook.com\n",
    "#PBS -j oe\n",
    "module load blast+/2.7.1\n",
    "blastp -db /panfs/panfs.cluster/opt/ncbi/db-static/db-11-01-18/nr -query {filename}{n} -out {filename}.blastout{n}  -evalue 0.000001 -outfmt 6 -max_target_seqs 50 -num_threads 12 -task blastp-fast\n",
    "'''\n",
    "for n in range(N):\n",
    "    open(qsubfile.format(n=n,filename=filename),'w').write(txt.format(n=n,filename=filename))\n",
    "\n",
    "# Annotate OGS2 with gene names\n",
    "#split files to 20\n",
    "from Bio import SeqIO\n",
    "filename = '/panfs/panfs.cluster/scratch/ks2073/Blast/OGS2/OGS2.fa'\n",
    "N = 20\n",
    "fouts = [open(filename+str(i),'w') for i in range(N)]\n",
    "for n, seq in enumerate(SeqIO.parse(filename,'fasta')):\n",
    "    towrite = '>'+seq.description + '\\n'+str(seq.seq)+'\\n'\n",
    "    fouts[n % N].write(towrite)\n",
    "for i in range(N):\n",
    "    fouts[i].close()\n",
    "# generate qsub scripts\n",
    "for n in range(N):\n",
    "    open(qsubfile.format(n=n,filename=filename),'w').write(txt.format(n=n,filename=filename))\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get gene info based on NCBI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "f_ogs2blast = r\"E:\\Lab\\works\\20121023_manduca_genome_new\\20120818Genome Ms\\OGS2201404\\GeneNaming\\20181129OGS2.fa.blastout\"\n",
    "f_mcot2blast = r\"E:\\Lab\\works\\20121023_manduca_genome_new\\20120818Genome Ms\\OGS2201404\\GeneNaming\\20181129MCOT.fasta.blastout\"\n",
    "import pandas as pd\n",
    "colnames = 'Query id, Subject id, % identity, alignment length, mismatches, gap openings, q. start, q. end, s. start, s. end, e-value, bit score'.split(', ')\n",
    "df = pd.read_csv(f_ogs2blast,sep = '\\t',header = None)\n",
    "df.columns = colnames\n",
    "df['gene'] = df['Query id'].apply(lambda x:x.split('-')[0])\n",
    "#filter, '% identity'>25, alignment length >50, e-value < 10^-6, bit score > 100\n",
    "df_ogs2blastFilter = df[(df['% identity'] > 25) & \n",
    "                                  (df['alignment length'] > 50) & \n",
    "                                  (df['e-value'] < 10**-6) & \n",
    "                                  (df['bit score'] > 100)]\n",
    "\n",
    "df = pd.read_csv(f_mcot2blast,sep = '\\t',header = None)\n",
    "df.columns = colnames\n",
    "df['gene'] = df['Query id'].apply(lambda x:'MCOT.' + x.split('.')[1])\n",
    "#filter, '% identity'>25, alignment length >50, e-value < 10^-6, bit score > 100\n",
    "df_mcot2blastFilter = df[(df['% identity'] > 25) & \n",
    "                                  (df['alignment length'] > 50) & \n",
    "                                  (df['e-value'] < 10**-6) & \n",
    "                                  (df['bit score'] > 100)]\n",
    "del df\n",
    "\n",
    "\n",
    "df_msex2blast = pd.concat([df_ogs2blastFilter,df_mcot2blastFilter])\n",
    "\n",
    "msex2Accs = set(df_msex2blast['Subject id'])\n",
    "#open gene accession file. store accessions that exist in df_ogs2blastFilter\n",
    "f_GeneAccession = r\"E:\\Insects\\ncbi_geneNames\\gene2accession\"\n",
    "f_GeneAccessionMsex = f_GeneAccession+'.Msex'\n",
    "fo = open(f_GeneAccession,'r')\n",
    "fout = open(f_GeneAccessionMsex,'w')\n",
    "line = fo.readline()\n",
    "fout.write(line)\n",
    "for line in fo:\n",
    "    accession = line.split('\\t')[5]\n",
    "    if accession in msex2Accs:\n",
    "        fout.write(line)\n",
    "fout.close()\n",
    "fo.close()\n",
    "\n",
    "#open gene accession file of Manduca OGS2, store in dataframe\n",
    "df_geneAcc = pd.read_csv(f_GeneAccessionMsex,sep = '\\t',dtype=str)\n",
    "#open gene name file, store genes that exist in df_geneAcc\n",
    "ogs2Gene = set(df_geneAcc['GeneID'])\n",
    "f_GeneName = r\"E:\\Insects\\ncbi_geneNames\\All_Data.gene_info\"\n",
    "fo = open(f_GeneName,'r')\n",
    "fout = open(f_GeneName+'.Msex','w')\n",
    "line = fo.readline()\n",
    "fout.write(line)\n",
    "for line in fo:\n",
    "    gene = line.split('\\t')[1]\n",
    "    if gene in ogs2Gene:\n",
    "        fout.write(line)\n",
    "fout.close()\n",
    "fo.close()\n",
    "\n",
    "#open gene name file of Manduca OGS2, store in dataframe\n",
    "df_geneName = pd.read_csv(f_GeneName+'.Msex',sep = '\\t',dtype=str)\n",
    "#combine df_geneAcc with df_geneName, get a dataframe with accession, geneID and geneName\n",
    "dftemp = pd.merge(df_geneAcc,df_geneName,on = 'GeneID',how = 'inner')\n",
    "df_accName = dftemp.loc[:,['GeneID','protein_accession.version','description']]\n",
    "df_accName = df_accName.drop_duplicates(subset = 'protein_accession.version')\n",
    "df_accName.columns = ['GeneID', 'Subject id', 'description']\n",
    "df_accName = df_accName[df_accName['description'].apply(lambda x:'uncharacterized' not in x)]\n",
    "\n",
    "\n",
    "#give each transcript in df_ogs2blastFilter a name\n",
    "dftemp = pd.merge(df_msex2blast,df_accName,on = 'Subject id',how = 'left')\n",
    "df_transcript = dftemp.iloc[:,[0,1,2,3,10,11,12,13,14]]\n",
    "df_transcriptSort = df_transcript.sort_values(by = ['Query id','bit score'], ascending =[True, False], axis=0)\n",
    "df_transcriptSortName = df_transcriptSort.dropna(axis = 0,subset = ['description'])\n",
    "df_transcriptBest = df_transcriptSortName.drop_duplicates(subset = 'Query id')\n",
    "\n",
    "#save df_transcriptBest to file\n",
    "df_transcriptBest.to_excel(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsMCOT_OGS2.xlsx\",index=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get GO term for each protein\n",
    "\n",
    "```python\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "file_ogs = r\"D:\\Insects\\ManducaSexta\\Interpro\\ms_ogs_proteins.tsv\"\n",
    "file_mcot = r\"D:\\Insects\\ManducaSexta\\Interpro\\Ms_MCOTprotein.tsv\"\n",
    "\n",
    "df_ogs = pd.read_csv(file_ogs, sep='\\t', header=None,low_memory=False)\n",
    "df_mcot = pd.read_csv(file_mcot, sep='\\t', header=None,low_memory=False)\n",
    "\n",
    "df_ogs = df_ogs[[0,13]]\n",
    "df_mcot = df_mcot[[0,13]]\n",
    "df_all = pd.concat([df_ogs, df_mcot], ignore_index=True)\n",
    "df_all.columns = ['ID','GO']\n",
    "\n",
    "df_all = df_all.dropna()\n",
    "\n",
    "df_go = df_all.groupby('ID')['GO'].apply(lambda x:'|'.join(x))\n",
    "df_go.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsMCOTOGS2_GO.csv\", sep='\\t')\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get KO term for each protein\n",
    "```python\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "use prevous result. Store file as `\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190922MsMCOT_OGS_KEGG_KO.txt\"`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connect OGS2 and MCOT with NCBI annotated manduca sexta\n",
    "20191107\n",
    "\n",
    "NCBI also annotated the genes for manduca sexta. The naming is better. connect MCOT, OGS2 and NCBI genes.\n",
    "\n",
    "download the protein sequences from `ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/262/585/GCF_000262585.1_Msex_1.0/GCF_000262585.1_Msex_1.0_protein.faa.gz`\n",
    "\n",
    "do the sequence comparison on server. the scripts looks like\n",
    "\n",
    "```bash\n",
    "python /cache/home/xc278/w/GitHub/MCuNovo/tools/SeqCompare.py -q /projectsp/f_jx76_1/xiaolong/temp/20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta -s /projectsp/f_jx76_1/xiaolong/temp/GCF_000262585.1_Msex_1.0_protein.faa -m 10 -e 0.02 -p /cache/home/xc278/p/anaconda3/bin/muscle -t 24 -o /projectsp/f_jx76_1/xiaolong/temp/20191107MsMCOT_NCBI.txt\n",
    "python /cache/home/xc278/w/GitHub/MCuNovo/tools/SeqCompare.py -q /projectsp/f_jx76_1/xiaolong/temp/ms_ogs_proteins.fa -s /projectsp/f_jx76_1/xiaolong/temp/GCF_000262585.1_Msex_1.0_protein.faa -m 10 -e 0.02 -p /cache/home/xc278/p/anaconda3/bin/muscle -t 24 -o /projectsp/f_jx76_1/xiaolong/temp/20191107MsOGS2_NCBI.txt\n",
    "```\n",
    "transfer the file to local. connect MCOT and OGS2 to NCBI IDs, and get the name\n",
    "```python\n",
    "# annotated sequence name to OGS2 and MCOT\n",
    "file_NCBI = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\GCF_000262585.1_Msex_1.0_protein.faa\"\n",
    "file_mcot2NCBI = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20191107MsMCOT_NCBI.txt\"\n",
    "file_ogs2NCBI = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20191107MsOGS2_NCBI.txt\"\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "df_mcot2ncbi = pd.read_csv(file_mcot2NCBI,sep='\\t')\n",
    "df_ogs2ncbi = pd.read_csv(file_ogs2NCBI, sep='\\t')\n",
    "# add MS column\n",
    "df_mcot2ncbi['MS'] = df_mcot2ncbi['match_len'] * df_mcot2ncbi['match_len'] / (df_mcot2ncbi['seq1_len'] * df_mcot2ncbi['seq2_len'])\n",
    "df_ogs2ncbi['MS'] = df_ogs2ncbi['match_len'] * df_ogs2ncbi['match_len'] / (df_ogs2ncbi['seq1_len'] * df_ogs2ncbi['seq2_len'])\n",
    "\n",
    "\n",
    "# sort by seq1_id and match_len\n",
    "df_mcot2ncbi =df_mcot2ncbi.sort_values(by=['seq1_id', 'match_len', 'MS'], ascending=[True, False, False])\n",
    "df_ogs2ncbi = df_ogs2ncbi.sort_values(by=['seq1_id', 'match_len', 'MS'], ascending=[True, False, False])\n",
    "\n",
    "# keep only the best match for each query\n",
    "df_mcot2ncbi = df_mcot2ncbi.drop_duplicates(subset = ['seq1_id'], keep='first')\n",
    "df_ogs2ncbi = df_ogs2ncbi.drop_duplicates(subset = ['seq1_id'], keep='first')\n",
    "\n",
    "# get gene names\n",
    "dc_names = {}\n",
    "for s in SeqIO.parse(file_NCBI,'fasta'):\n",
    "    seq_id = s.id\n",
    "    name = s.description.split(' ',1)[1].replace('[Manduca sexta]', '')\n",
    "    dc_names[seq_id] = name\n",
    "\n",
    "df_mcot2ncbi['NCBI_annotationName'] = df_mcot2ncbi['seq2_id'].apply(lambda x:dc_names[x])\n",
    "df_ogs2ncbi['NCBI_annotationName'] = df_ogs2ncbi['seq2_id'].apply(lambda x:dc_names[x])\n",
    "\n",
    "df_all = pd.concat([df_mcot2ncbi, df_ogs2ncbi])\n",
    "#save result\n",
    "excel = pd.ExcelWriter(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20191107MsOGSMCOT2_toNCBIannotatedNames.xlsx\", engine='xlsxwriter')\n",
    "df_mcot2ncbi.to_excel(excel,sheet_name='mcot2NCBI',index=None)\n",
    "df_ogs2ncbi.to_excel(excel,sheet_name='OGS2NCBI',index=None)\n",
    "df_all.to_excel(excel,sheet_name='mcotOGS2NCBI',index=None)\n",
    "excel.save()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### interproscan for NCBI Ms proteins\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "install interproscan, signalP, panther and TMHMM as instructed. change the settings to fit the running need.\n",
    "\n",
    "Run interproscan for NCBI generated models for Manduca sexta on Perceval.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --job-name=interpro\n",
    "#SBATCH --partition=main\n",
    "#SBATCH --time=7-00:00:00\n",
    "#SBATCH --error=err.%x.%j.%N\n",
    "#SBATCH --output=out.%x.%j.%N\n",
    "#SBATCH --export=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=24\n",
    "#SBATCH --mem=120GB\n",
    "#SBATCH --mail-user=xiaolong.cao@rutgers.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --requeue                    # Return job to the queue if preempted\n",
    "\n",
    "cd /projectsp/f_jx76_1/xiaolong/temp\n",
    "module load java/11.0.5\n",
    "/projectsp/f_jx76_1/xiaolong/p/interpro/interproscan-5.38-76.0/interproscan.sh  -d ./  -goterms -iprlookup -pa -f TSV,XML,HTML -i GCF_000262585.1_Msex_1.0_protein.faa --cpu 24 \n",
    "\n",
    "```\n",
    "\n",
    "Also do the same search for MCOT and OGS2 to use newer database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## process and filter data\n",
    "20190209"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filter the proteins and add protein names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " set up files for analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* fMemBand: cell membrane of hemocytes and fat body; each gel band\n",
    "* fMemGroup: cell membrane of hemocytes and fat body; each group\n",
    "* fDevBand: developmental stages of hemolymph; each gel band\n",
    "* fDevGroup: developmental stages of hemolymph; each group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the original output of maxquant, revoe contaminant protein and “rev__” proteins. \n",
    "Also remove proteins with total unique peptide count < 1. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "```python\n",
    "fMemBand = r\"D:\\Insects\\ManducaSexta\\2017Ms_fatbody_hemocyte_membrane\\raw_32group\\combined\\txt\\proteinGroups.txt\"\n",
    "fMemGroup = r\"D:\\Insects\\ManducaSexta\\2017Ms_fatbody_hemocyte_membrane\\raw_32group\\combined\\txt_4groupWIthPi\\proteinGroups.txt\"\n",
    "fDevBand = r\"D:\\Insects\\ManducaSexta\\1520\\txt_band\\proteinGroups_band_develop.txt\"\n",
    "fDevGroup = r\"D:\\Insects\\ManducaSexta\\1520\\txt_group\\proteinGroups_group_develop.txt\"\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "dfMemBand = pd.read_csv(fMemBand, sep=\"\\t\",low_memory=False)\n",
    "dfMemGroup = pd.read_csv(fMemGroup, sep = \"\\t\", low_memory=False)\n",
    "dfDevBand = pd.read_csv(fDevBand, sep = \"\\t\", low_memory=False)\n",
    "dfDevGroup = pd.read_csv(fDevGroup, sep = \"\\t\", low_memory=False)\n",
    "\n",
    "print(\"dfMemBand\",dfMemBand.shape)\n",
    "print(\"dfMemGroup\", dfMemGroup.shape)\n",
    "print(\"dfDevBand\", dfDevBand.shape)\n",
    "print(\"dfDevGroup\", dfDevGroup.shape)\n",
    "\n",
    "# remove last 14 columns\n",
    "dfMemBand = dfMemBand.iloc[:,:(dfMemBand.shape[1] - 14)]\n",
    "dfMemGroup = dfMemGroup.iloc[:,:(dfMemGroup.shape[1] - 14)]\n",
    "dfDevBand = dfDevBand.iloc[:,:(dfDevBand.shape[1] - 14)]\n",
    "dfDevGroup = dfDevGroup.iloc[:,:(dfDevGroup.shape[1] - 14)]\n",
    "\n",
    "# remove comtaminant proteins\n",
    "def removeContaminant(df):\n",
    "    '''\n",
    "    return a dataframe contaminant proteins and revert proteins were removed\n",
    "    '''\n",
    "    return df[df['Protein IDs'].apply(lambda x: x[:5] != \"CON__\" and x[:5] != \"REV__\")]\n",
    "\n",
    "dfMemBand = removeContaminant(dfMemBand)\n",
    "dfMemGroup = removeContaminant(dfMemGroup)\n",
    "dfDevBand = removeContaminant(dfDevBand)\n",
    "dfDevGroup = removeContaminant(dfDevGroup)\n",
    "\n",
    "print(\"dfMemBand\",dfMemBand.shape)\n",
    "print(\"dfMemGroup\", dfMemGroup.shape)\n",
    "print(\"dfDevBand\", dfDevBand.shape)\n",
    "print(\"dfDevGroup\", dfDevGroup.shape)\n",
    "\n",
    "# remove proteins with < 1 unique peptides\n",
    "def removeRareProtein(df):\n",
    "    '''\n",
    "    return a dataframe that proteins with <1 unique peptides were removed\n",
    "    '''\n",
    "    return df[df['Unique peptides']>0]\n",
    "\n",
    "dfMemBand = removeRareProtein(dfMemBand)\n",
    "dfMemGroup = removeRareProtein(dfMemGroup)\n",
    "dfDevBand = removeRareProtein(dfDevBand)\n",
    "dfDevGroup = removeRareProtein(dfDevGroup)\n",
    "\n",
    "print(\"dfMemBand\",dfMemBand.shape)\n",
    "print(\"dfMemGroup\", dfMemGroup.shape)\n",
    "print(\"dfDevBand\", dfDevBand.shape)\n",
    "print(\"dfDevGroup\", dfDevGroup.shape)\n",
    "\n",
    "# save files\n",
    "folder = r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane'\n",
    "dfMemBand.to_excel(os.path.join(folder,'membrane','20190921MsMemBand_ori.xlsx'))\n",
    "dfMemGroup.to_excel(os.path.join(folder,'membrane','20190921MsMemGroup_ori.xlsx'))\n",
    "dfDevBand.to_excel(os.path.join(folder,'developHemolymph',\"20190921MsDevBand_ori.xlsx\"))\n",
    "dfDevGroup.to_excel(os.path.join(folder,'developHemolymph',\"20190921MsDevGroup_ori.xlsx\"))\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "original dataframe shape\n",
    "```\n",
    "dfMemBand (4419, 332)\n",
    "dfMemGroup (4419, 80)\n",
    "dfDevBand (1887, 906)\n",
    "dfDevGroup (1887, 150)\n",
    "```\n",
    "after remove the last 14 columns and contaminant and reverse proteins\n",
    "```\n",
    "dfMemBand (4295, 318)\n",
    "dfMemGroup (4295, 66)\n",
    "dfDevBand (1824, 892)\n",
    "dfDevGroup (1824, 136)\n",
    "```\n",
    "\n",
    "after remove proteins with <1 unique peptides\n",
    "```\n",
    "dfMemBand (4238, 318)\n",
    "dfMemGroup (4238, 66)\n",
    "dfDevBand (1824, 892)\n",
    "dfDevGroup (1824, 136)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check and verify the relationship between iBAQ and intensity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read in files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "file = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\membrane\\20190921MsMemBand_ori.xlsx\"\n",
    "df = pd.read_excel(file,sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calculate iBAQ/intensity for each sample\n",
    "\n",
    "The conclusion is, yes, iBAQ = intensity/peptides, where peptides is the number of theoretical detectable peptides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iBAQ</th>\n",
       "      <th>iBAQ FP1</th>\n",
       "      <th>iBAQ FP2</th>\n",
       "      <th>iBAQ FP3</th>\n",
       "      <th>iBAQ FP4</th>\n",
       "      <th>iBAQ FP5</th>\n",
       "      <th>iBAQ FP6</th>\n",
       "      <th>iBAQ FP7</th>\n",
       "      <th>iBAQ FP8</th>\n",
       "      <th>iBAQ FS1</th>\n",
       "      <th>...</th>\n",
       "      <th>iBAQ HP7</th>\n",
       "      <th>iBAQ HP8</th>\n",
       "      <th>iBAQ HS1</th>\n",
       "      <th>iBAQ HS2</th>\n",
       "      <th>iBAQ HS3</th>\n",
       "      <th>iBAQ HS4</th>\n",
       "      <th>iBAQ HS5</th>\n",
       "      <th>iBAQ HS6</th>\n",
       "      <th>iBAQ HS7</th>\n",
       "      <th>iBAQ HS8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>761</th>\n",
       "      <td>23.999618</td>\n",
       "      <td>24.000542</td>\n",
       "      <td>23.999798</td>\n",
       "      <td>24.000684</td>\n",
       "      <td>23.999811</td>\n",
       "      <td>23.999377</td>\n",
       "      <td>24.000173</td>\n",
       "      <td>23.999599</td>\n",
       "      <td>23.998618</td>\n",
       "      <td>23.999786</td>\n",
       "      <td>...</td>\n",
       "      <td>23.999706</td>\n",
       "      <td>23.999674</td>\n",
       "      <td>23.998810</td>\n",
       "      <td>24.001292</td>\n",
       "      <td>24.000517</td>\n",
       "      <td>23.998952</td>\n",
       "      <td>23.999585</td>\n",
       "      <td>24.000308</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>24.000492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>23.000372</td>\n",
       "      <td>22.999888</td>\n",
       "      <td>22.999952</td>\n",
       "      <td>22.999826</td>\n",
       "      <td>22.999673</td>\n",
       "      <td>22.999594</td>\n",
       "      <td>23.000211</td>\n",
       "      <td>23.000869</td>\n",
       "      <td>23.000700</td>\n",
       "      <td>22.999765</td>\n",
       "      <td>...</td>\n",
       "      <td>23.000035</td>\n",
       "      <td>22.999799</td>\n",
       "      <td>23.000547</td>\n",
       "      <td>22.999833</td>\n",
       "      <td>23.000158</td>\n",
       "      <td>22.999784</td>\n",
       "      <td>22.999762</td>\n",
       "      <td>22.999518</td>\n",
       "      <td>22.999242</td>\n",
       "      <td>23.000115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034</th>\n",
       "      <td>32.001135</td>\n",
       "      <td>31.999482</td>\n",
       "      <td>31.998986</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>32.000208</td>\n",
       "      <td>31.999432</td>\n",
       "      <td>31.999120</td>\n",
       "      <td>31.998705</td>\n",
       "      <td>32.000192</td>\n",
       "      <td>31.999444</td>\n",
       "      <td>...</td>\n",
       "      <td>31.999569</td>\n",
       "      <td>32.000143</td>\n",
       "      <td>31.999694</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>31.999305</td>\n",
       "      <td>31.999810</td>\n",
       "      <td>32.001481</td>\n",
       "      <td>32.000427</td>\n",
       "      <td>31.999621</td>\n",
       "      <td>31.999165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3695</th>\n",
       "      <td>20.000319</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>19.999205</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000422</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>19.999781</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>19.999470</td>\n",
       "      <td>19.999660</td>\n",
       "      <td>19.999239</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000652</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000429</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           iBAQ   iBAQ FP1   iBAQ FP2   iBAQ FP3   iBAQ FP4   iBAQ FP5  \\\n",
       "761   23.999618  24.000542  23.999798  24.000684  23.999811  23.999377   \n",
       "950   23.000372  22.999888  22.999952  22.999826  22.999673  22.999594   \n",
       "2034  32.001135  31.999482  31.998986  32.000000  32.000208  31.999432   \n",
       "3695  20.000319  20.000000  19.999205  20.000000  20.000422  20.000000   \n",
       "\n",
       "       iBAQ FP6   iBAQ FP7   iBAQ FP8   iBAQ FS1    ...       iBAQ HP7  \\\n",
       "761   24.000173  23.999599  23.998618  23.999786    ...      23.999706   \n",
       "950   23.000211  23.000869  23.000700  22.999765    ...      23.000035   \n",
       "2034  31.999120  31.998705  32.000192  31.999444    ...      31.999569   \n",
       "3695  20.000000  19.999781  20.000000  20.000000    ...      19.999470   \n",
       "\n",
       "       iBAQ HP8   iBAQ HS1   iBAQ HS2   iBAQ HS3   iBAQ HS4   iBAQ HS5  \\\n",
       "761   23.999674  23.998810  24.001292  24.000517  23.998952  23.999585   \n",
       "950   22.999799  23.000547  22.999833  23.000158  22.999784  22.999762   \n",
       "2034  32.000143  31.999694  32.000000  31.999305  31.999810  32.001481   \n",
       "3695  19.999660  19.999239  20.000000  20.000652  20.000000  20.000429   \n",
       "\n",
       "       iBAQ HS6   iBAQ HS7   iBAQ HS8  \n",
       "761   24.000308  24.000000  24.000492  \n",
       "950   22.999518  22.999242  23.000115  \n",
       "2034  32.000427  31.999621  31.999165  \n",
       "3695  20.000000  20.000000  20.000000  \n",
       "\n",
       "[4 rows x 33 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract iBAQ and intensity\n",
    "df_iBAQ = df.loc[:,[e for e in df.columns if 'iBAQ' in e]]#iBAQ columns\n",
    "df_intensity = df.loc[:,[e for e in df.columns if 'Intensity' in e]]#intensity columns\n",
    "df_ratio = df_iBAQ.copy()\n",
    "for i in range(df_ratio.shape[1]):\n",
    "    df_ratio.iloc[:,i] = df_intensity.iloc[:,i] / df_iBAQ.iloc[:,i] #intensity/iBAQ\n",
    "df_ratioGood = df_ratio.dropna()\n",
    "df_ratioGood.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iBAQ</th>\n",
       "      <th>iBAQ FP1</th>\n",
       "      <th>iBAQ FP2</th>\n",
       "      <th>iBAQ FP3</th>\n",
       "      <th>iBAQ FP4</th>\n",
       "      <th>iBAQ FP5</th>\n",
       "      <th>iBAQ FP6</th>\n",
       "      <th>iBAQ FP7</th>\n",
       "      <th>iBAQ FP8</th>\n",
       "      <th>iBAQ FS1</th>\n",
       "      <th>...</th>\n",
       "      <th>iBAQ HP7</th>\n",
       "      <th>iBAQ HP8</th>\n",
       "      <th>iBAQ HS1</th>\n",
       "      <th>iBAQ HS2</th>\n",
       "      <th>iBAQ HS3</th>\n",
       "      <th>iBAQ HS4</th>\n",
       "      <th>iBAQ HS5</th>\n",
       "      <th>iBAQ HS6</th>\n",
       "      <th>iBAQ HS7</th>\n",
       "      <th>iBAQ HS8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>iBAQ</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP7</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FP8</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS7</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ FS8</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP7</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HP8</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS7</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iBAQ HS8</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          iBAQ  iBAQ FP1  iBAQ FP2  iBAQ FP3  iBAQ FP4  iBAQ FP5  iBAQ FP6  \\\n",
       "iBAQ       1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP1   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP2   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP3   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP4   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP5   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP6   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP7   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP8   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS1   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS2   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS3   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS4   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS5   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS6   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS7   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS8   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP1   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP2   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP3   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP4   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP5   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP6   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP7   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP8   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS1   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS2   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS3   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS4   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS5   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS6   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS7   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS8   1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "          iBAQ FP7  iBAQ FP8  iBAQ FS1    ...     iBAQ HP7  iBAQ HP8  \\\n",
       "iBAQ           1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP1       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP2       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP3       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP4       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP5       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP6       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP7       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FP8       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS1       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS2       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS3       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS4       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS5       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS6       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS7       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ FS8       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP1       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP2       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP3       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP4       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP5       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP6       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP7       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HP8       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS1       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS2       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS3       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS4       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS5       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS6       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS7       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "iBAQ HS8       1.0       1.0       1.0    ...          1.0       1.0   \n",
       "\n",
       "          iBAQ HS1  iBAQ HS2  iBAQ HS3  iBAQ HS4  iBAQ HS5  iBAQ HS6  \\\n",
       "iBAQ           1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP1       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP2       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP3       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP4       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP5       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP6       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP7       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FP8       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS1       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS2       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS3       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS4       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS5       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS6       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS7       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ FS8       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP1       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP2       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP3       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP4       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP5       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP6       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP7       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HP8       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS1       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS2       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS3       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS4       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS5       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS6       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS7       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "iBAQ HS8       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "          iBAQ HS7  iBAQ HS8  \n",
       "iBAQ           1.0       1.0  \n",
       "iBAQ FP1       1.0       1.0  \n",
       "iBAQ FP2       1.0       1.0  \n",
       "iBAQ FP3       1.0       1.0  \n",
       "iBAQ FP4       1.0       1.0  \n",
       "iBAQ FP5       1.0       1.0  \n",
       "iBAQ FP6       1.0       1.0  \n",
       "iBAQ FP7       1.0       1.0  \n",
       "iBAQ FP8       1.0       1.0  \n",
       "iBAQ FS1       1.0       1.0  \n",
       "iBAQ FS2       1.0       1.0  \n",
       "iBAQ FS3       1.0       1.0  \n",
       "iBAQ FS4       1.0       1.0  \n",
       "iBAQ FS5       1.0       1.0  \n",
       "iBAQ FS6       1.0       1.0  \n",
       "iBAQ FS7       1.0       1.0  \n",
       "iBAQ FS8       1.0       1.0  \n",
       "iBAQ HP1       1.0       1.0  \n",
       "iBAQ HP2       1.0       1.0  \n",
       "iBAQ HP3       1.0       1.0  \n",
       "iBAQ HP4       1.0       1.0  \n",
       "iBAQ HP5       1.0       1.0  \n",
       "iBAQ HP6       1.0       1.0  \n",
       "iBAQ HP7       1.0       1.0  \n",
       "iBAQ HP8       1.0       1.0  \n",
       "iBAQ HS1       1.0       1.0  \n",
       "iBAQ HS2       1.0       1.0  \n",
       "iBAQ HS3       1.0       1.0  \n",
       "iBAQ HS4       1.0       1.0  \n",
       "iBAQ HS5       1.0       1.0  \n",
       "iBAQ HS6       1.0       1.0  \n",
       "iBAQ HS7       1.0       1.0  \n",
       "iBAQ HS8       1.0       1.0  \n",
       "\n",
       "[33 rows x 33 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ratioGood.mean(axis=1)\n",
    "df_ratioGood.corr()#correlation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check intensity and iBAQ values of group and band"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read in files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "%matplotlib inline\n",
    "\n",
    "folder = r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\membrane'\n",
    "file_g = os.path.join(folder,\"20181125MsMemGroup_ori.xlsx\")\n",
    "file_b = os.path.join(folder, \"20190921MsMemGroup_ori.xlsx\")\n",
    "df_g = pd.read_excel(file_g,sheet_name='Sheet1')\n",
    "df_b = pd.read_excel(file_b,sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check if all indexes the same. Answer is Yes.\n",
    "for band and group analysis, the indexes are the same.  \n",
    "They should be the same as they were processed together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,j in zip(df_g.index, df_b.index):\n",
    "    if i!=j:\n",
    "        print(i,j)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get working columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g_iBAQ = df_g.loc[:,[e for e in df_g.columns if 'iBAQ' in e]]#iBAQ columns\n",
    "df_b_iBAQ = df_b.loc[:,[e for e in df_b.columns if 'iBAQ' in e]]#iBAQ columns\n",
    "df_g_intensity = df_g.loc[:,[e for e in df_g.columns if 'Intensity' in e]]#iintensity columns\n",
    "df_b_intensity = df_b.loc[:,[e for e in df_b.columns if 'Intensity' in e]]#intensity columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check if iBAQ/Intensity sum values are the same in band and group. Answer is Yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(df_b_iBAQ['iBAQ'] == df_g_iBAQ['iBAQ']) == df_b_iBAQ.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(df_b_intensity['Intensity'] == df_g_intensity['Intensity']) == df_b_intensity.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check if adding bands together equals group\n",
    "Answer is Yes. plot is sum of iBAQ in bands vs iBAQ in group analysis. most values are close to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F3507D68>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F708B4A8>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F70B9B00>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F70A71D0>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F701C860>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F701C898>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEVCAYAAADpbDJPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3X+4XFV97/H354IECyoJgZgm6AGbqtC0iikEbeu5ohCgGixgY3kkQSwq8FjvTVtD8RbEquC9/kKrGJUChQpUbImK5kb06NVCSELBBGlMwAinBBDCr4BFAt/7x1pDdiZ7fpyTc+bH2Z/X88xzZtZes/Z39lkz39lrr9lbEYGZmVXPf+t2AGZm1h1OAGZmFeUEYGZWUU4AZmYV5QRgZlZRTgBmZhXlBGBmVlFOAD1E0u2SBrsdh9lYcH/ufU4APSQiDomIIUnnSXpa0tZ8u0PSCfX1JR0o6VlJny9ZJkl/JWmDpF9JulvSRyXt0ZlXY1VX688A3erTkoYk/VdhvVslHZGXhaQnctl/SvqkpN3GcBP0PCeA3nV1ROwdEXsD7weukDStrs4pwMPAAkmT6pZdBJye67wAOAZ4I3DV+IZt1lC3+vRZtfXm242FZb+X4zkS+DPgz0f1yvqUE0APkbRJ0hvryyNiOfA48LK6RacAHwSeBt5caGcWcAZwckTcGBHbIuJ24ATgOEmvH6/XYFbTqD9D7/XpiPgP4P8Bv7Mr7fQbJ4Ael3d7jwP2AH5aKP9DYCbp2881pDdOzZHAcETcXGwrIu4BbgKOGu+4zRrpxT4t6WDgD4F/35V2+o0TQO96m6RHgCeAZcBHI+KRwvKFwLcj4mHgn4BjJO2fl00FNjdodzOw3zjFbNZMt/r0RZIeybdb6pbdIulh4BvAl4F/GNlL6m9OAL3rmojYJyJ+g7SbfIqkdwNIej5wEnAlQB7TvJs0hgnwIDC9QbvTgV+OZ+BmDXSrT78vr3efiDi0btmhETE5Il4WER+MiGdH99L6kxNAH4iITcC32T4m+lbghcDnJd0n6T5gBtt3mb8HHCDpsGI7kg4A5gI/6ETcZo24T/cGJ4A+IGkmMA+4PRctBC4BZgOvyrfXAa+SNDsifgZcDFwpaa6k3SQdAlwL/Bvw3U6/BrMi9+ne4ATQu/60Nm8ZWAX8GPiQpBmkA2Kfjoj7Crc1wHdIbySAs0hjmlcATwLrgF8Ax1dtN9d6hvt0j5GvCFYNks4Hjgf+qO7Am1lfcp/edU4AFSLpLGBjRHyn27GYjQX36V3jBGBmVlE+BmBmVlG7dzuAZqZOnRoDAwOly5544gn22muvzgbUg7wdkmbbYc2aNQ9GRN/8+K1Zv++EfupT/RJrp+Nst8/3dAIYGBhg9erVpcuGhoYYHBzsbEA9yNshabYdJP2is9Hsmmb9vhP6qU/1S6ydjrPdPt/2EFCed/vvkr6ZHx8oaWU+NevVtVOySpqUH2/MywcKbZydy9dLOnpkL8nMzMbSSPYA/gK4g/RrPYALgU9FxFWSLgZOA76Q/z4cEb8laUGu96f5ZEsLgEOA3wS+K+m3I+KZMXotlbT2Px9l0ZJvjeg5my44bpyiMRt/A4X+vnj2trb7v/v9ztraA8i/2juO9CMMJAl4A/C1XOUy0nxcgPn5MXn5kbn+fOCqiHgqIn4ObAR2+Fm3mZl1Trt7AJ8G/pp0EQaAfYFHImJbfjxMOm8H+e89ABGxTdKjuf4M0mlbKXnOcySdTrroA9OmTWNoaKg0oK1btzZcViXTnp++BY3ERNxu7g9mI9cyAUj6Y+CBiFij7df3VEnVaLGs2XO2F0QsBZYCzJkzJxodOOmXgz/j7bNXXscn1o7sWP6mkwfHJ5gucn8wG7l2PjleB7xF0rHAnqRjAJ8G9pG0e94LmAncm+sPAwcAw5J2B14EbCmU1xSfY2ZmHdbyGEBEnB0RMyNigHQQ93sRcTLwfeDEXG0hcF2+v4ztJ286MdePXL4gzxI6EJgF7HB1HzMz65xd+R3AB4CrJP0d6TJqX8nlXwH+UdJG0jf/BQARcbuka0iXgNsGnOkZQGZm3TOiBBARQ8BQvn8XJbN4IuK/SFf2KXv+R4CPjDRIMzMbez4XkJlZRTkBmJlVlBOAmVlFOQGYmVWUE4CZWUU5AZiZVZQTgJlZRTkBmJlVlBOAmVlFOQGYmVWUE4CZWUU5AZiZVZQTgJlZRTkBmJlVlBOAmVlFOQGYlZB0iaQHJK0rlE2RtELShvx3ci6XpIskbZT0E0mHFp6zMNffIGlh2brMusUJwKzcpcC8urIlwA0RMQu4IT8GOIZ0idNZwOnAFyAlDOBc4HDSxZPOrSUNs17gBGBWIiJ+SLqkadF84LJ8/zLg+EL55ZHcBOwjaTpwNLAiIrZExMPACnZOKmZdsyvXBDarmmkRsRkgIjZL2j+XzwDuKdQbzmWNynci6XTS3gPTpk1jaGhobCMfga1bt3Z1/a0snr3tufvTnr/j42a8TXfWMgFIOgC4HHgx8CywNCI+k3dvrwYGgE3A2yLiYUkCPgMcCzwJLIqIW3JbC4EP5qb/LiIuw6z/qaQsmpTvXBixFFgKMGfOnBgcHByz4EZqaGiIbq6/lUVLvvXc/cWzt/GJte19j9108uA4RdRar27TdoaAtgGLI+KVwFzgTEkH4/FQq57789AO+e8DuXwYOKBQbyZwb5Nys57QMgFExObaN/iIeBy4g7Qb6/FQq5plQG0mz0LgukL5KXk20Fzg0TxUtBw4StLk/GXnqFxm1hNGdAxA0gDwamAl4zQe2u5YaK+OqXXaSMZAaybidhvr/iDpq8AgMFXSMGnv9QLgGkmnAXcDJ+Xq15OGPDeShj1PBYiILZI+DKzK9c6PiPoDy2Zd03YCkLQ3cC3w/oh4LA31l1ctKWt7PLTdsdBeHVPrtM9eeV3bY6A13RwLHS9j3R8i4u0NFh1ZUjeAMxu0cwlwyZgFZjaG2poGKul5pA//KyPi67nY46FmZn2sZQLIs3q+AtwREZ8sLPJ4qJlZH2tn7OB1wDuAtZJuzWV/g8dDzcz6WssEEBE/onz8HjweambWt3wqCDOzinICMDOrKCcAM7OKcgIwM6soJwAzs4pyAjAzqygnADOzivIFYcysawYK5/a3zvMegJlZRTkBmJlVlBOAmVlFOQGYmVWUE4CZWUU5AZiZVZSngZpZJYxmyummC44bh0h6h/cAzMwqygnAzKyiPARkZmPCv+rtP94DMDOrqI4nAEnzJK2XtFHSkk6v36zT3OetV3V0CEjSbsDfA28ChoFVkpZFxE87GYdZp/Rrn/dwTjLRZw51+hjAYcDGiLgLQNJVwHygp98MZrvAfb5iypLG4tnbWNQkmXQraXQ6AcwA7ik8HgYOL1aQdDpwen64VdL6Bm1NBR4c8wj7z4i3gy4cp0i6q9l2eGknA6nTss/DiPp9J/TNe+t9fRJrqzjH4T3ZVp/vdAJQSVns8CBiKbC0ZUPS6oiYM1aB9Stvh6SHt0PLPg/t9/tO6OFtuZN+ibVX4+z0QeBh4IDC45nAvR2OwayT3OetZ3U6AawCZkk6UNIewAJgWYdj6DhJt0sa7HYc1hXu89azOpoAImIbcBawHLgDuCYibh9lcz2xu9yOiDgkIoYAJJ0n6WlJW/PtDkkn1D8nf2A8K+nzJcsk6a8kbQB+V9Ldkj6aP2BKSRqS9F+F9W6VdEReNl/SrZIek/SgpBskDYzV6++QnuwPY9znO2WXt+V49nlJv6r1eeArjWLIff5ddWWDkobr6tTeFw9K+rqk6bvy2hvoyf5JRPjWwRtwHnBF4fHRwK+AaXX1zgUeArYAk+qWfRbYABxBOo5zCHAz8PUm6x0C3lVS/lvAo8CRpPHqFwAnAC/p9rbybWLceqnPA4PAcFkdYArwPeCqbm+zTt38S+AOkLRJ0hvLlkXEcuBx4GV1i04BPgg8Dby50NYs4Azg5Ii4MSK2RfpGeQJwnKTXjzC8VwE/j4gbInk8Iq6NiLtH2I7Zc3q8z5eKiC3AtcDvjEV7/cAJoIvybu1xwB4U5oVL+kPSwcKrgGtIb4yaI0nfYG4uthUR9wA3AUeNMIxbgFdI+pSk/y5p75G/ErP29EifbxTbVFJS+fexaK8f9HwCaPUzekmLJP0yj2HfWj/m16PeJukR4AnSAcGPRsQjheULgW9HxMPAPwF/nF/jOtJ84s3FxvKb6iLgUOA9kg5tsN6LJD2Sb7cARPqB0iBpvvo1wIOSLu3FRCDpEkkP5O1QtnxQ0qOFvvC3nY6xF0iaImlFHi9fIWlyg3oLc50NkhYWyl8jaW1+z10kSc3arfW/XP8nxf4n6RngPcCfSXqa7X3+18Dqwnu62Of/mdTnfy5pJWmYcnNu7+y8nvWSjs7l+zX5nPispG2SnpH0a+CbuZ1Jkq4m/Sbji5IeA27L7f3PXKfVZ88kSVfn5StVOG5WEifN2lQ6/rEyb9urlY/njfvnW7fHoJrdgN2AO4GDSN8YbgMOrquzCPhct2Nt8To2AW/M989jx/HQAWAd8O78+PmkMfmTCnX+E/g/ud57gF/UtX8s8G3gB8A/ACtLYhii5BhASb3fB+4CPtbt7VYS2x+Rkty6BssHgW92O85u34CPA0vy/SXAhSV1puT/8xRgcr4/OS+7mTTWrtyvjmnWbqH/CZgLrKz1eWBrsc/n9/QvgPWkYZ3bgFcX+3wufxR4P2nW1Or8nINz/UnAgfmz4QfAR8s+J3KfXwUsyO1eDHySNDX3jPx4CPgicHXd9mnns+cM4OJ8f0GtjQZx7tasTdKXr2Kc7833FzGOn2+9vgfw3M/oI+LXpN3D+V2OaUxFxCbSm6c25vlW4IXA5yXdJ+k+YF9gXl7+PeAASYcVmplP+mYzF/gqsI9GOZMhIlYBX6cHx0Ej4oekA4TW3Hzgsnz/MuD4kjpHAysiYkukb90rgHm537ww0lh7AJcXnt+o3fnA5ZHcBOxD+rArcxjwH6S9gGNJ7+kPUOjzwKeAPUnDQF8jfWAeAJxJOkD7VET8nPRBfgRwP40/J16R26jF/Aclr2UNcGRtT6cQZ6vPnmIbXyu0Mb8uzo25vdI283PeUBdn2f9szPV6Aij7Gf2Mknon5F3Pr0k6oGR5z5I0k/ThXpsauBC4BJhNOkD7KuB1pG8VkyLiZ6RvCFdKmqt0srGXk94c/wZ8l8bbqWz9fyDpzyXtnx+/AngLaWy1Hx0h6TZJ35Z0SLeD6ZJpEbEZIP/dv6ROo/fWjHy/vrxZu2Vt1RLAnqRTXBwt6fhc9yG29/lh4LXs2Od/TvoAfBXwSuBh4FLgZGAPSbvl/+3BwM9IP6wrey3PA34VaSpurXy/kpifJe1x7NvG9qGsTl5HrY1m27asfF/gkbo4i+sat8+3Xk8A7fyM/hvAQET8LunD77Kdn9Jz/lR5TjRpF/XHwIckzSAd8Pp0RNxXuK0h7erWxnLPAr4MXAE8CbweuA84PiKezXV2Ot1AA4+QPvDX5ni+A/wLaXe/39wCvDQifo80bfBfuxzPuJH0XUnrSm7t7iE3em+1deqKNtqqeQlpDvw+pD3LK0kf7j8GPkTq0zMp9HngGWAtqS/Wjk38Nemg8XtJfX4daXip0Vl0GsVcK2/1OtvZDiPdhqPZ5uP6+dbrVwRr+TP6iHio8PBLNO4QXRMRA4X755HGRMs8SeP/yankA1j5Q/7CfEPSGtKUulpHKttOgw1iW0dhyl0/i4jHCvevl/R5SVMjoudPFjZSEVE6xRJA0v2SpkfE5jyk80BJtWHSMZOamaTx8OF8v1he60uN2i17n76mtrdA6u/nSbqU9I399RHxnhzrXsA5EbG2LrYDIuJYSbuTxsEfIn0YEhEfk3Q+8D9IX3y2lqz/XuB9wC8l7Z6/Xc8ENkTE0ZKW53UM5nVcwI7Di+2cwqNWZzi38aLcRrPnlpU/SBq2LcZ5b36t4/r51ut7AC1/Rl831v0W0q8tq+ZvSW+CuZLmAo8W3nyVIenFhRkrh5H690PNnzUhLWP7N+eFwHUldZYDR0manGfzHAUsz/3m8Ty8KNI4fO35jdpdBpySJgNt73+57Unw3BTL15E+xFudGqO4nhOB7+XjEcuABbnNfwCeIg3zlH5O5Od8P7dRFnPZOmraOYVHyzglHQjMIh1YH3Gc4/75Nl5Hl8fqRjpQ9DPS0fNzctn5wFvy/Y+RxhJvyxvxFd2OeRy2wVdJ09OeJn27OI00G+g9eblIFx25k7TrPKfbMXdpO5xV6As3Aa/tdsxd2k77AjeQfjl7AzAll88Bvlyo907SAcqNwKmF8jmkIZY7gc8BatFuaf8jje2vzf+PtcBpubzVe3pP0lTQjaQPzoMKsZ2Tn7eePDupUZu5/KDcxsbc5qRW62jWZhfiHNfPt9o/1szMKqbXh4DMzGyc9PRB4KlTp8bAwEDX1v/EE0+w1157dW39u6JfYx+PuNesWfNgROzXumZv6Ha/b0c/9K8qx9hun+/pBDAwMMDq1au7tv6hoSEGBwe7tv5d0a+xj0fckn4xpg2Os273+3b0Q/+qcozt9nkPAZmZVVRP7wFYawNLvlVavnj2NhY1WLbpguPGMySzcdWoz9erfw+43+/MewBmZhXlBGBmVlFOAGZmFeUEYGZWUU4AZnUkHSDp+5LukHS7pL/I5aO5IlbpVbfMeoETgNnOtgGLI+KVpIvsnCnpYNJVsG6IiFmkc+HULul3DOmEX7NI577/AqSEAZxLuuzgYcC5anB5RrNucAIwqxMRmyOids3kx0lnYJzBCK6Ilc/iWHrVrQ6+FLOm/DsAsyaULvT9atJ1bne4IlbtKmqM/ApQZes5nbT3wLRp0xgaGhqz1zAetm7d2rUYF8/e1roSMO35O9btxW3aze0ITgBmDUnaG7gWeH9EPLbjJWN3rFpSNqKra0XEUtKVs5gzZ05U9RQG7Wj0A8d6i2dv4xNrt3/EbTp5cJwiGr1un67CQ0BmJSQ9j/Thf2VEfD0X31+7QEcbV8S6t0m5WU9wAjCrk6+E9RXgjoj4ZGHRiK6IRYOrbnXkRZi1wUNAZjt7HfAOYK2kW3PZ35CuG3uNpNOAu4GT8rLrSVd62ki6rvOpABGxRdKHSZcCBDg/IorXnTXrKicAszoR8SPKx+8BjiypH8CZDdq6BLhk7KIzGzseAjIzqygnADOzinICMDOrKCcAM7OKcgIwM6soJwAzs4pyAjAzqygnADOzinICMDOrqJYJQNIlkh6QtK5Q5isjmZn1uXb2AC5l54tY+MpIZmZ9rmUCiIgfAvUnsPKVkczM+txoTwZXiSsjdftqPe1odHWk+qshFfXya+qHbW42UYz12UAn1JWRun21nnY0ujpS/dWQinrxykg1/bDNzSaK0c4C8pWRzMz63GgTgK+MZGbW59qZBvpV4Ebg5ZKG89WQLgDeJGkD8Kb8GNKVke4iXRnpS8AZkK6MBNSujLQKXxnJepynP1sVtDwGEBFvb7DIV0ayiexS4HPA5YWy2vTnCyQtyY8/wI7Tnw8nTX8+vDD9eQ7pmNcaScvyTDizrvMvgc1KePqzVYGvCWzWvkpMf25HN6frNpreXK9+KnQvbtNuT3t2AjDbdRNq+nM7ujldt9HU53r1U6F7cfpzt6c9ewjIrH2e/mwTihOAWfs8/dkmFA8BmZXI058HgamShkmzeS4ArslToe8GTsrVrweOJU1/fhI4FdL0Z0m16c/g6c/WY5wAzEp4+rNVgYeAzMwqygnAzKyinADMzCrKCcDMrKKcAMzMKsoJwMysopwAzMwqyr8DMLNKGGjzHEJFmy44bhwi6R3eAzAzqyjvAfSQ0XxDMTMbLe8BmJlVlPcAKmi0exoTfTzUrGqcAMysazzs2V0eAjIzqygnADOziur4EJCkecBngN2AL0fEBZ2OoRMm4q6t51GPTlX6/EQ00ft8RxOApN2AvwfeRLpe6ipJyyLip52Mw6xTqtTnJ+KXnomu03sAhwEbI+IuAElXAfOBnn0zuFPvmpFuv8WztzE4PqF0S9/1eWj9f1s8exuL/N4oNZI+X9uO3dprULqaXYdWJp0IzIuId+XH7wAOj4izCnVOB07PD18OrO9YgDubCjzYxfXvin6NfTzifmlE7DfGbbalnT6fy2v9fiqwL93t9+3oh/5V5Rjb6vOd3gNQSdkOGSgilgJLOxNOc5JWR8ScbscxGv0ae7/G3UTLPg/b+31+/QPjHtUu6of/k2NsrdOzgIaBAwqPZwL3djgGs05yn7ee1ekEsAqYJelASXsAC4BlHY7BrJPc561ndTQBRMQ24CxgOXAHcE1E3N7JGEZol4eiJN0uaXAMYhmpnhhGG4V+jbvUKPp8v7z+fojTMbbQ0YPAVSfpPOAc4KlcdA/wwYi4tq7egcCdwMURcUbdMgF/STpgOBP4JXAFcF5E/LrBeoeAKyLiy4WywVw2Mz8+BPgU8Pukces7gf8VEdeP/hWbWS/zL4E77+qI2Dsi9gbeD1whaVpdnVOAh4EFkibVLbuI9OF/CvAC4BjgjcBVuxjXN4AVwDRgf+B9wGO72KaZ9TAngHEmaZOkN5Yti4jlwOPAy+oWnQJ8EHgaeHOhrVnAGcDJEXFjRGzLwwknAMdJev0oY5wKHAh8KSJ+nW8/jogfjaY9M+sTETFhb8AU0rfaDfnv5Ab1FuY6G4CFhfLXAGuBjaRv3mrWLmno5KJc/yfAocAm0jf0Z4D7gC2kg4ACjgMeAfbJz58H3E2aJvgh4LPAskI8ZwFP5PZXAgOFZXflttcDRxfK5wFPAg8ASwrlC0hDURuAq/PfbwInAv9ato4m23leXu/G4joKyyfldZTFfXYuL4t7pzZJiWplIe49cvki0nDYrfn2ror0uVcAN+b/5V+283/pQow7vS8KbT1T+J8t65G+1HAdE62/d/1DepzfjB+vbUxgCXBhgzfsXfnv5Hy/1nFvBo7IHfjbwDHN2gWOzfUEzM3/uE2kBLAVOA/4NelD/8nc+f86P3c30rj71cB1wG3A20l7AfvnOsuA+/P9BaThJICDSUNGX84d5s7cXq3Nmwrreyyv/2ngofz8i3PH/FzuVAH8kDREdXWLbVxbx0HAHjnug+vqnEE6nlEW9235DVMW905tAtcACwpxv7fwhvhcBfvc/qTjNh+hkABabMOuvy8K69nag32pdB09FuMixqC/d/1DepzfjOuB6fn+dGB9SZ23A18sPP5iLpsO/EdZvUbt1p5bt/572DEBXFFYPgCsA95NekOtAB4FTiJ9IJ+dO8b7c/3bgc35/u6kXxAq17sL+Ehetjy3d0S+PwS8q9Cm8nqGc/0jgOWF584nvWFvrK2jyTZ+7rn58dnA2XV1lgNHNIj77Pp6jdrMz3kQ2L0k7jF5Q/RbnyvUPY8dE0DD/0unY6T8fVGrt7WdmDvcl0rX0WMxLmIM+vtEPwYwLSI2A+S/+5fUmUH6kK4ZzmUz8v368mbtlrW1W76/J+ng7dGSjs/P3UT6oH1zfu7uwAuBzwMfIB0HmEE6JgDwPGCapMMiTS98lHTagJeTfmz0g5LXUPba9iUlpLLXNgNYQzqB2e8U1tFIo3WU1qmLu9m2bxT3I7mNsnWdIOknkr4mqfjjq07qdJ9rpNn/pRfeF7Xn7ClptaSbgOOb1NsprnHsS43W0eq1dDJGGIP+3vdXBJP0XeDFJYvOabeJkrJoUj7StmpeQkoArwI+LWktadx2HnB9fu4s4BJS7CcCryYlg1WSZpOGbf4RuDKfUwbgt0kzgTYA362LtSzBl722FwH7SfqtvGwy8E7S0NFBLV53O9tppNu43biL6/oG8NWIeErSe4DLgDeURryLeqzPNfK3wExJc/PjFwG/IWlFm88f7/dF7TkviYh7JR1E6m/fb9F2J/rSaJd1MsYx6e99nwAionSGDYCk+yVNj4jNkqaTDoTWG4YdTkA5kzRkMpzvF8trP+Fv1G7Zz/6fyXHem6bwUzvt3zrSWPx1pAO+RwK/CXw6Iu6T9ALSWSTXSPoO6YDcMGl3+qekuf8HAT/ObV0bEc+WxFp2GoIHgb1JM5AA9gO2kRJI7fX/X9IxgFtIB5cbaedUB7U6w5J2J30YbWnx3EZx7yNp9/yt6Ln6EfFQof6XgAubxLxLeqzPNfJu0m9Djs5xnZ1jv65H3he1/1vt712SVpH2ZsvaLsY1rn2pyTqavpZOxjhm/X1Xx5B6+Qb8b3Y8KPXxkjpTgJ+TvvVOzven5GWrSAetage7jm3WLunDvXiw6+ZcPhmYlO9PJX1brz9wtDtpHP9Ath8IOqSuzpnseHDpGuB80pjqWrYfXLqLNPTUsE3gn9nx4NIZjdbRYhuPKu58/xB2PCi2K3FPL6zvrcBNVehzhTbPY8djAM22YS+/L+7pgb7U8D3QrL1+7O9d/5Ae5zfjvsANuWPdUOjAc0hXZqrVeydp+tVG4NRC+RzSt+s7STNk1KJdkcbO7yR9IM/J5a/Nj2/Lf09rEO+xwM/y88/JZecDb8n398wdYiNpJsZBufws4NL8vPXkWRmN2szlB+U2NuY2JzVbR4vtPKq487Jzxijuj5EOkt9GGkZ4RUX63ItJ3yxrs7uGgRe22IY9+77okb7U9D3QIzGOSX/3qSDMzCpqos8CMjOzBnr6IPDUqVNjYGCg22HwxBNPsNdee3U7jKYcY2Nr1qx5MLp0RTCzXtbTCWBgYIDVq1d3OwyGhoYYHBzsdhhNOcbGJP2i4ys16wMeAjIzq6iWewD5F2aXk2YbPAssjYjPSJpCOm/NAOl8N2+LiIfz+eo/Qzqq/SSwKCJuyW0tJP26FeDvIuKysX051TOw5FsALJ69jUX5fiubLjiudSUz62mSTiJN/30lcFhEjHi4pJ09gG3A4oh4JWkO75mSDibN870hImaRpnwtyfWPIf2idRbpl69fyMFOAc4FDgcOA86VNHmkAZuZVY2kQUmX1hWvA/6EdOLGUWmZACJic+0bfEQ8Trqs3QzSCcNq3+AvI53Hg1x+eSQ3kX7JNh04GlgREVsi4mHSic/mjTZwM7Mqi4g7ImL9rrQxooPAkgZI56dZSd2JnyS1OvFTOydRQtLppD2VqZi0AAAI00lEQVQHpk2bxtDQ0EhCHBdbt27tiTjKLJ6dzhM17fnb77fSrdfSy9vRrIraTgCS9gauJZ2a+LF8XpvSqiVlrU5stL0gYin5Qslz5syJXpjZ0sszbBYVjgF8Ym17/85NJw+OY0SN9fJ2NOtFklaSTh2xNzBF0q150QciXVFwl7T1iSHpeaQP/ysj4uu5eKQnfmp0cikzMysREYdDOgZAmlCzaCzbb3kMIM/q+QpwR0R8srBoGekMleS/1xXKT1EyF3g0DxUtB46SNDkf/D0ql5mZWRe0swfwOuAdwNrC7sffABcA10g6jXQd25PysutJU0A3kqaBngoQEVskfZh0JkGA8yOi2WmGzcysAUlvJV03fD/gW5JujXz673a1TAAR8SMaX+jkyJL6QTodallbl5AueGJmZm2KiCHqhswj4l+Af9mVdv1LYDOzinICMDOrKCcAM7OKcgIwM6soJwAzs4pyAjAzqygnADOzinICMDOrKCcAM7OKcgIwM6soJwAzs4pyAjAzqygnADOzinICMDOrKCcAM7OKcgIwM6soJwAzs4pyAjAzq6h2Lgp/iaQHJK0rlE2RtELShvx3ci6XpIskbZT0E0mHFp6zMNffIGlh2brMzKxz2tkDuBSYV1e2BLghImYBN+THAMcAs/LtdOALkBIGcC5wOHAYcG4taZiZWXe0TAAR8UNgS13xfOCyfP8y4PhC+eWR3ATsI2k6cDSwIiK2RMTDwAp2TipmZtZBoz0GMC0iNgPkv/vn8hnAPYV6w7msUbmZmXXJ7mPcnkrKokn5zg1Ip5OGj5g2bRpDQ0NjFtxobd26tSfiKLN49jYApj1/+/1WuvVaenk7mlXRaBPA/ZKmR8TmPMTzQC4fBg4o1JsJ3JvLB+vKh8oajoilwFKAOXPmxODgYFm1jhoaGqIX4iizaMm3gPTh/4m17f07N508OI4RNdbL29GsikY7BLQMqM3kWQhcVyg/Jc8Gmgs8moeIlgNHSZqcD/4elcvMzKxLWn5llPRV0rf3qZKGSbN5LgCukXQacDdwUq5+PXAssBF4EjgVICK2SPowsCrXOz8i6g8sm5lZB7VMABHx9gaLjiypG8CZDdq5BLhkRNGZmdm48S+BzcwqygnAzKyinADMzCrKCcDMrKKcAMzMKsoJwMysopwAzMwqygnAzKyinADMzCpqrM8GartgIJ/YzcysE7wHYGZWUU4AZmYV5QRgZlZRTgBmZhXlBGBmVlFOAGZmFeUEYGZWUf4dQAWN9vcGmy44bowjMbNucgIYJ/5Rl5n1uo4PAUmaJ2m9pI2SlnR6/WZmlnR0D0DSbsDfA28ChoFVkpZFxE87GYeNzmj2ajxsZNa7Oj0EdBiwMSLuApB0FTAf6EgCGO2wzOLZ21jkIZ1RKW7zdrejk4ZZZygiOrcy6URgXkS8Kz9+B3B4RJxVqHM6cHp++HJgfccCbGwq8GC3g2jBMTb20ojYrwvrNetpnd4DUEnZDhkoIpYCSzsTTnskrY6IOd2OoxnHaGYj1emDwMPAAYXHM4F7OxyDmZnR+QSwCpgl6UBJewALgGUdjsHMzOjwEFBEbJN0FrAc2A24JCJu72QMo9RTQ1INOEYzG5GOHgQ2M7Pe4XMBmZlVlBOAmVlFTfgEIGmKpBWSNuS/kxvUW5jrbJC0sFD+Gklr86krLpKkZu0quSjX/4mkQwttPSPp1nxb1uq0GJImSbo6L18paaCw7Oxcvl7S0YXy0jbzgfeVOd6r80H4puto1l6HY1wk6ZeFbfeusv+hmY1QREzoG/BxYEm+vwS4sKTOFOCu/Hdyvj85L7sZOIL0G4ZvA8c0axc4NtcTMBdYWVjP1sL93YA7gYOAPYDbgIPr4joDuDjfXwBcne8fnOtPAg7M7ezWrE3gGmBBvn8x8N5m6+ixGBcBn+t2X/LNt4l2m/B7AKRTTVyW718GHF9S52hgRURsiYiHgRXAPEnTgRdGxI0REcDlhec3anc+cHkkNwH75HbqPXdajIj4NVA7LUaj2L8GHJn3QOYDV0XEUxHxc2Bjbq+0zfycN+Q2yuItW0cvxWhm46AKCWBaRGwGyH/3L6kzA7in8Hg4l83I9+vLm7XbqC2APSWtlnQT6cOtUb2d4oqIbcCjwL4t4i0r3xd4JLdRv65G62j1WjoZI8AJeUjta5KKPyY0s1GaENcDkPRd4MUli85pt4mSsmhSPpq2AF4SEfdKOgi4Cfh+i7ZHGldZQm/1Oka7rJMxfgP4akQ8Jek9pL2DN5TUN7MRmBAJICLe2GiZpPslTY+IzXko5oGSasPAYOHxTGAol8+sK6+duqJRuw1PdxERtb93SVpFOtldWdvFuA4AhiXtDrwI2NJsHQ3KHyQNRe2ev2EX6zdaR9PX0skYI+KhQv0vARdiZrusCkNAy4DarJ6FwHUldZYDR0manGfzHAUsz0M7j0uam8eoTyk8v1G7y4BT8mygucCjOUlMljQJQNJU4LeBfdX8tBjFdZwIfC8fi1gGLMgzcA4EZpEOVpeeaiM/5/u5jbJ4y9ZBo/Y6HWPdMZS3AHdgZruu20ehx/tGGlu+AdiQ/07J5XOALxfqvZN0oHIjcGqhfA6wjjRz5XNs//V0o3ZFuujNncBaYE4uf21+fFv+exppxtDPct1zcr3zgbfk+3sC/5xjuhk4qBDXOfl568kzk3L5Tm3m8oNyGxtzm5NaraNRe12I8WPA7XnbfR94Rbf7lW++TYSbTwVhZlZRVRgCMjOzEk4AZmYV5QRgZlZRTgBmZhXlBGATgqSTJN0u6VlJvu6wWRucAKzvSBqUdGld8TrgT4Afdj4is/40IX4JbBYRdwBsP4+dmbXiPQAzs4ryHoD1DUkrSdcX2BuYIunWvOgDEbG8e5GZ9ScnAOsbEXE4pGMAwKKIWNTVgMz6nIeAzMwqygnAJgRJb5U0TLp857ckeUjIrAWfDM7MrKK8B2BmVlFOAGZmFeUEYGZWUU4AZmYV5QRgZlZRTgBmZhXlBGBmVlH/H0J01mF3sYUYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_bg_iBAQ = df_g_iBAQ.copy()\n",
    "for column in df_bg_iBAQ.columns[1:]:\n",
    "    df_bg_iBAQ[column] = df_b_iBAQ.loc[:,[column+str(e+1) for e in range(8)]].sum(axis=1)\n",
    "(df_bg_iBAQ / df_g_iBAQ).hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F329AA20>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F32EEA90>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F740E160>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F34647F0>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F2347E80>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x000001C9F2347EB8>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEVCAYAAADpbDJPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3X20XFWd5vHvIy/BJiIJkQCB9oLGRpDuUQIBnXbdAeW1F7FbwDAICcJkUFi2a7JGo0wPNMoyMNpqGluMQhN84UW0mygwdARuI7a8D69izCVEuJ1IhEAgoQUiv/lj70pO6lbdW/fm1ts9z2etWlW1z6599jm16/zO2WfXOYoIzMysfN7Q7gqYmVl7OACYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlANAiUi6TNLftLseZmPFbXrbOAA0kaRVkj7QQL4+SWc1uz4RcXZEfD7Ps1fSQLPnaeNHo+055+26Ni3pAkmvSdpQeHw6T+uT9Puc9qykH0nac6yWo10cAMzMtrg2IiYWHpcUpp0bEROBdwC7Al9pTxXHjgNAC0iaK+lOSV+S9LykJyUdm6ddBPw5cGneu7g0p+8vaZmkdZKWSzq5UN6Vkr4u6UZJL0m6W9Lb8jRJ+oqktZLWS3pY0rsKn/uCpJ2Bm4G9Cns6e0l6WdJuhfkcLOl3knZo3dqyblDmNh0R64AfAu8abRmdwgGgdWYCy4EpwCXA5ZIUEecBPyPvXUTEubkxLwO+D+wOnAL8g6QDC+WdAvwtMAnoBy7K6UcB72fLXspHgOeKFYmIjcCxwOrCns5qoA84uZD1o8A1EfHaGK0DG19K2aYlTQE+DPy/0ZbRKRwAWuc3EfGtiPgDsATYE5haJ+9fAKsi4h8jYlNEPEDa4zixkOdHEXFPRGwCvgf8p5z+GvAmYH9AEfF4RKxpsI5LSD8QJG1H+kF+p/FFtJIZj236ZEkvFB57FaYtkvQC8BCwBvgfDdahYzkAtM5vKy8i4uX8cmKdvG8FZhYbInAqsEet8oCXK2VFxG3ApcDXgWckLZa0S4N1vAE4QNJ+wAeB9RFxT4OftfIZj236uojYtfBYXZj2yZw2LSJOjYjfNViHjuUA0BmqL8n6NPCvVQ1xYkR8vKHCIhZFxMHAgaTD5v/ZwDyJiN8D15F+mKfhvX8bPbfpLuAA0BmeAfYrvP8J8A5Jp0naIT8OkfTO4QrK+Wbmk1wbgd8Df6gzz90kvbkq/SpgLnAC8N1RLIsZuE13BQeAzvA14MQ8mmJRRLxEOvE1G1hNOjS+GJjQQFm7AN8Cngd+QzpZ9qXqTBHxK+BqYGWxrzMifg68DjwQEau2dcGstNymu4B8QxirJuk24PsR8e1218VsLLhN1+YAYFuRdAhpuN4+ea/NrKu5TdfnLiDbTNIS4KfAp/xDsfHAbXpoPgIwMyspHwGYmZXU9u2uwFCmTJkSPT09Nadt3LiRnXfeubUV6kBeD8lQ6+H+++9/NiLe0uIqjdpQ7X6sjZf2Mx6WYyyXodE239EBoKenh/vuu6/mtL6+Pnp7e1tboQ7k9ZAMtR4k/aa1tdk2Q7X7sTZe2s94WI6xXIZG23zDXUCStpP0/yT9JL/fN1+xb4WkayXtmNMn5Pf9eXpPoYzP5vTlko4e2SKZmdlYGskRwF8Dj5P+lAHpTxxfiYhrJF0GnAl8Iz8/HxFvlzQ75/uIpANIfwI5ENgL+Kmkd+QLSdkoPfLv65m74MYRfWbVwuObVBuz5uup0d7nH7Rp2N+B2/1gDR0BSNobOB74dn4v4Ajg+pxlCfCh/HpWfk+efmTOP4t0GdZXIuJJ0uVeDx2LhTAzs5Fr9Ajgq8CnSZdkBdgNeCFfthVgAJiWX08jXfiJiNgkaX3OPw24q1Bm8TObSZoHzAOYOnUqfX19NSu0YcOGutPKZOob097PSIzH9eb2YDZywwYASX8BrI2I+yX1VpJrZI1hpg31mS0JEYuBxQAzZsyIeidFxsNJn7Hw99+7gS8/MrJz+atO7W1OZdrI7cFs5BrZcrwPOEHSccBOpHMAXwV2lbR9PgrYm3SBJ0h79vsAA5K2B94MrCukVxQ/Y2ZmLTbsOYCI+GxE7B0RPaSTuLdFxKnA7Wy5m88c0o0XAJbm9+Tpt0X6u/FSYHYeJbQvMB3wzUbMzNpkW/4H8BngGklfIN0b8/KcfjnwHUn9pD3/2QAR8Zik64BfApuAczwCyMysfUYUACKij3STZSJiJTVG8eQ78JxU5/MXseVGz2Zm1ka+FpCZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmNUg6QpJayU9WkibLGmZpBX5eVJOl6RFkvolPSzpPYXPzMn5V0iaU2teZu3iAGBW25XAMVVpC4BbI2I6cGt+D3As6Ran04F5wDcgBQzgfGAm6eZJ51eChlkncAAwqyEi7iDd0rRoFrAkv14CfKiQflUkdwG7StoTOBpYFhHrIuJ5YBmDg4pZ22zLPYHNymZqRKwBiIg1knbP6dOApwv5BnJavfRBJM0jHT0wdepU+vr6xrbmdWzYsKFl8xor8w/aNCht6htrpxd1+nK247sYNgBI2ge4CtgDeB1YHBFfy4e31wI9wCrg5Ih4XpKArwHHAS8DcyPigVzWHOB/5aK/EBFLMOt+qpEWQ6QPToxYDCwGmDFjRvT29o5Z5YbS19dHq+Y1VuYuuHFQ2vyDNvHlR4benK06tbdJNRob7fguGukC2gTMj4h3AocB50g6APeHWvk8k7t2yM9rc/oAsE8h397A6iHSzTrCsAEgItZU9uAj4iXgcdJhrPtDrWyWApWRPHOAGwrpp+fRQIcB63NX0S3AUZIm5Z2do3KaWUcY0TkAST3Au4G7aVJ/aKN9od3Yd9kMjfR9VhuP622s24Okq4FeYIqkAdLR60LgOklnAk8BJ+XsN5G6PPtJ3Z5nAETEOkmfB+7N+S6MiOoTy2Zt03AAkDQR+CHwqYh4MXX1185aI63h/tBG+0K7se+yGf7+ezcM2/dZrdP7QkdjrNtDRJxSZ9KRNfIGcE6dcq4ArhizipmNoYaGgUragbTx/15E/Cgnuz/UzKyLDRsA8qiey4HHI+LvCpPcH2pm1sUa6Tt4H3Aa8IikB3Pa53B/qJlZVxs2AETEndTuvwf3h5qZdS1fCsLMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykfFN4M2ubnhr397XW8RGAmVlJOQCYmZWUu4DMrBRG0920auHxTahJ5/ARgJlZSfkIwMzGhE/odh8fAZiZlZQDgJlZSbW8C0jSMcDXgO2Ab0fEwlbXwayV3Oa713g/cdzSACBpO+DrwAeBAeBeSUsj4petrIdZq3Rrm3d//uh1U9Bo9RHAoUB/RKwEkHQNMAvo6B+D2TZoa5sfbmM0/6BNzPXGvrRaHQCmAU8X3g8AM4sZJM0D5uW3GyQtr1PWFODZMa9h9xnxetDFTapJew21Ht7ayopUGbbNw4ja/Zj65Dj5HXX7cuTf5FguQ0NtvtUBQDXSYqs3EYuBxcMWJN0XETPGqmLdyush6eD1MGybh8bb/Vjr4PU2IuNhOdqxDK0eBTQA7FN4vzewusV1MGslt3nrWK0OAPcC0yXtK2lHYDawtMV1GLckXSbpb9pdD9uK23yTuL1vu5YGgIjYBJwL3AI8DlwXEY+NsriWHy4XSVol6QMN5u2TdFaTqrJ5PUTE2RHx+TzPXkkDoy1U0gWSXpO0ofD4dJ52oKR/kfS8pBck3S/puG1flG3S1vZQzxi3+WZoaL11UHvfrNjegb4xaO/frZEekt5eyFP5Tbwg6d8kHT7aedbQ8jbc8v8BRMRNwE1jUE5H/uBbrcnr4dqI+GiN9B8D3wD+Ir8/hNp93S3Tye1hrNp8M3Tyehuhn5COrprt2oj4qKQdgIuAH0naKyIGndcZqXZ8F/4n8BiQNFfSnZK+lPeKn5R0bJ52EfDnwKV5z+HSnL6/pGWS1klaLunkQnlXSvq6pBslvSTpbklvy9Mk6SuS1kpaL+lhSe8qfO4LknYGbgb2Kuy97yXpZUm7FeZzsKTf5cbc6LJOAfYFvhURr+bHzyPizm1fk9YNytTe64mI14AlwB7AbsNk71gOAGNnJrCcNJTrEuBySYqI84CfAedGxMSIODc32GXA94HdgVOAf5B0YKG8U4C/BSYB/aS9DYCjgPcD7wB2BT4CPFesSERsBI4FVud5ToyI1UAfcHIh60eBa3JjbtRzuT7flfQhSVNH8FkbP8rS3muSNAGYCwxERNcOPyUiOvoBHENqaP3AghrT5wK/Ax7Mj7NaVK9VwAcKdegvTPsj0lC/PfL7vmK9SI34Z1XlfRM4P7++knTJgMq0fwE2AY8CRwC/Bg4D3pCnC1gEvAg8A7wH6CU1Tqrm+/P8ejvgt8ChdZbvAuBV4IXCY688bW/gUuAJ4HXgDmB6C9b5FcBa4NE603uB9YW28L/b3X5bsE4mkzauK/LzpDr55uQ8K4A5hfSDgUfy72sRoDrlPgV8ILe1ZbltPJzbWqW9/yGv9w3Agw209+/k3/aLwD2FaccBvwImALfneT0G7FvI89n8ueeAoyvtna23F98vtPe35XKeAq4FdszpE/L7dbktv1jV5gN4e43fxFrgNtL5naG2T5Xy+4G7gZ6qZejPnz+6kF5zm0c68r47fyfFZZjLKLeBbW/AwzTu7Ugbmf2AHYGHgAOq8swFLm1D3VaxdQC4s2p6seH0sXUA+DSDN64bgG/k6VcCXyjk/yRpw/5o4f39+UtfDHyYdAh8JfCPuZH0MjgA7AQ8n9fnMcDyIZbvAuC7DayHffK8f9GCdf5+0gZnqADwk3a32xa3w0sqGwlgAXBxjTyTgZX5eVJ+PSlPuwc4nLRhvxk4tk6560kB4DjShv9O0k7I3YX2vnGE7X19botX5fZ9QOF7HAA+AVyW2/tK4Pe5vR9C2hZcBfw9aRtxRP5McXvxcGEed5COEMhlfjy/rszjAtKRy7VV6646AHy3MK2R7dMngMvy69mV8oEDcv4JpA37E7m8umUC1wGzayzDXEa5Dez0LqDNf6OPiFeByt/ou031CaKngX+NiF0Lj4kR8fE6n3+YtHeSCotYFBEHAweSDo0/R/oxAPw76VB50qBKRPye1IhOBU4j7YFtk4h4mnStm3dta1kNzOsO0p6abTGL1BdNfv5QjTxHA8siYl1EPE/agz9G0p7ALhHxi0hbkqsKn68u948K6T8HiIi7gF1zOUXDtnfStZHuinSJjNdJe/jVv+1ZwJKIWERq5xvz88WkbcHrpA18P/AnpA1mcXtxNelI4lTStuSyGuupuJxPAUdKanRAQyPbp2L51xfKn0XqjnolIp7My3BovTLzZ47IZVQvw6h1egCo9Tf6aTXyfTifHLpe0j41prfbM6SIXvET4B2STpO0Q34cIumdwxWU883MJ7I2kvaK3sTg9fQGYDdJb64q4irSHsMJwKBhbw3Mf5Kkv5X0dklvyCeFPwbcNdKymuRwSQ9Jurmqj3m8mhoRawDy8+418tT7HU3Lr6vTa5Vb2VZMY+sgXPzMBEn3AQeRjjArBrV30tHcS4U8LzH4tz0NmCxpJukIZT2pm2lijeXZAXgzqVuzmP40qX3uyJadpGKdi+sm8jwaPanbyPZpc55IQ4Ir5Q/1ndRK3w14IZdRa16j2gZ2egBo5G/0Pyb1q/0p8FO2RNtO8jXgxDxiYlFEvEQ6uTWb9K/Q35L2aiY0UNYuwLdIXTm/IfWBrqyRbxVpD2il0pjlvQAi4uekPacHImLVKJblVaCHtK5fJJ2XeIUUVNrtAeCtEfFnpK6Bf25zfcaEpJ9KerTGo9Gj4Xq/o4YuU9FgWQDvj3Qpg/9O2mtdP0R7/yiDtz/V8xZpY19p728lBZ8Ha9Th30lHJh8ptndgTS73lar2XpnXaNZBsX7DfXak634039Xot4Ht7sMcpn/zcOCWwvvPAp8dIv92wPp217tJ66KH+n3f3wROKbxfDuw5RFm30aKT5a1cDzXyrgKmtLvOTV4fm79rYE9qnNchjbD5ZnV7yfl/VStfvXIbbWuk81EnDlHvYX/bpD/PHZ5fb0+6UJqq81by1Sszt/eXgO2r511vHg2u+1Yug/JnBy1D1fxGtA3s9COAYf9GX9X/eALp35ZlsxQ4PY+ZPozUANbUyijpENKJ1GtbWcFWkLRHpf9W0qGkPcznhv5U11tKGuFDfr6hRp5bgKNy990k0t74LbmNvCTpsLzeTi98vl65NdtaLnsCbP6vyPsY+pLXjVwio1iHE4HbIm3llgKzJU2QtC8wnXQyu1aZK0nt/ae5jFrLU2sejWjVMizNn7m91jJs0zaw3XswDUTZ40jDHp8AzstpFwIn5NdfJJ1AeiivoP3bXecmrIOrSYeyr5H6/s4EzgbOztNFOhH7BGlI34w65Swh9UHObfcyNWk9nFtoC3cB7213nVuwTnYDbiUNDbwVmJzTZ7D1UOKPkU409gNnFNJnkLrxniAN7dUw5dZsa8B78/uH8vOZDdR9uN/2TsAPcp3vAfYrfPa8/Lnl5JFLNcp8oNLeSefg7sll/QCYMNw8Glz/zV6G8wrp9ZZh1NvAypdtZmYl0+ldQGZm1iQtvxjcSEyZMiV6enqaUvbGjRvZeeedm1J2s7nuI3P//fc/GxFvaelMt0Ez23093dCmuqGO0Bn1bLTNd3QA6Onp4b777mtK2X19ffT29jal7GZz3UdG0m9aOsNt1Mx2X083tKluqCN0Rj0bbfPuAjIzK6mOPgIom54FNzaUb/5Bm5ib865aeHwzq2TWVI22edjS7t3mx46PAMzMSsoBwMyspBwAzMxKygHAzKykHADMqkjaR9Ltkh6X9Jikv87pk5Xua7siP0/K6ZK0SFJ/viTvewplzcn5V0iaU2+eZu3gAGA22CZgfkS8k3TXq3MkHUC6M9atETGddH2cBTn/saSLeU0H5gHfgBQwgPNJ9889FDi/EjTMOoEDgFmViFgTEQ/k1y+Rrq44jfp335oFXBVJ8S5ZNe/E1cJFMRuS/wdgNgRJPcC7SfdZ3uouWZIqd98a6d2das1nHunogalTp9LX1zdmy9CIDRs2tHyekMb2N2rqG1P+dtRzJNq1LkfDAcCsDkkTgR8Cn4qIF4e4Vew233ErIhaTbnjOjBkzotWXEmjX5QvmjvCPYF9+ZHtWndrbvAqNgU64FESj3AVkVkO+b+0Pge9FxI9y8jOVm2/k57U5fQAo3od1b9KtD+ulm3UEBwCzKvnuWJcDj0fE3xUmjeguWdS5E1dLFsKsAe4CMhvsfcBpwCOSKjcg/xywELhO0pnAU8BJedpNpLs49QMvA2cARMQ6SZ8n3eYP4MKIWNeaRTAbngOAWZWIuJPa/fcAR9bIH8A5dcq6Arhi7GpnNnbcBWRmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUsMGAElXSFor6dFCmu+MZGbW5Ro5AriSwTex8J2RzMy63LABICLuAKovYOU7I5mZdbnRXgyu6++M1Il37Wn07kiVOyMBHbcMw+nE9W5WVmN9NdCuuTNSJ961p9G7I1XujAR0/N2RqnXiejcrq9GOAvKdkczMutxoA4DvjGRm1uUaGQZ6NfAL4E8kDeS7IS0EPihpBfDB/B7SnZFWku6M9C3gE5DujARU7ox0L74zknU4D3+2Mhj2HEBEnFJnku+MZOPZlcClwFWFtMrw54WSFuT3n2Hr4c8zScOfZxaGP88gnfO6X9LSPBLOrO38T2CzGjz82crA9wQ2a1zXD3+up13Dcxsd+gxbhj93+jDibhrq7ABgtu26ZvhzPe0antvo0GfYMvy504c+d9NQZ3cBmTXOw59tXHEAMGuchz/buOIuILMa8vDnXmCKpAHSaJ6FwHV5KPRTwEk5+03AcaThzy8DZ0Aa/iypMvwZPPzZOowDgFkNHv5sZeAuIDOzknIAMDMrKQcAM7OScgAwMyspBwAzs5LyKCAz6yo9I/j3cNGqhcePcU26n48AzMxKykcAXW40e0PeEzIz8BGAmVlpOQCYmZWUA4CZWUk5AJiZlZRPAjfJaIeqmZm1igOAmY0J7/R0HweAEvIfaayMPGR6sJafA5B0jKTlkvolLWj1/M1azW3eOlVLjwAkbQd8Hfgg6X6p90paGhG/bGU9RsKHtbYturHNW3m0ugvoUKA/IlYCSLoGmAX4x9AFfAg9Kl3Z5r3jk4xmPVx5zM5NqElztDoATAOeLrwfAGYWM0iaB8zLbzdIWt6kukwBnm1S2U31yS6quy4elNSOur+1xfMrGrbNQ0vbfT0d36a6pd3/l4s7op4NtflWBwDVSIut3kQsBhY3vSLSfRExo9nzaQbXvasM2+ahde2+nm74XrqhjtA99YTWnwQeAPYpvN8bWN3iOpi1ktu8daxWB4B7gemS9pW0IzAbWNriOoxLki6T9DftrocN4jZvHaulASAiNgHnArcAjwPXRcRjraxDwagPtyWtkvSBBvP2STprtPOqY1DdI+LsiPh8nmevpIHRFi7pAknfrZEekt6eX+8q6QpJv5X0kqRfS/rMaOo+nnVYmx9KN3wv3VBH6J56tv6PYBFxE3BTq+dbox5d8yVV65C6fwXYGXgnsB54B/Cu4T7UIXVvqU5p80Pphu+lG+oI3VNP8MXgtpmkuZLulPQlSc9LelLSsXnaRcCfA5dK2iDp0py+v6RlktblPwidXCjvSklfl3Rj3rO+W9Lb8jRJ+oqktZLWS3pY0rsKn/uCpJ2Bm4G98jw3SNpL0suSdivM52BJv5O0wygX/RDg+xHxfES8HhG/iojrR1mWmbWBA8DYmAksJw1TuwS4XJIi4jzgZ8C5ETExIs7NG+hlwPeB3YFTgH+QdGChvFOAvwUmAf3ARTn9KOD9pL3tXYGPAM8VKxIRG4FjgdV5nhMjYjXQB5xcyPpR4JqIeG2Uy3wXcJGkMyRNH2UZZtZOEdF1D2AyaSO6Ij9PqpNvTs6zAphTSD8YeIS0cV0EaKhySUP5FuX8D5NGcXwgT3sdeAV4kHRy749Iw/z2yNP7gLOAY0hB4rfAk1X1/CZwIXAt8CKwFujJ047L7/tJ48kHgMNIwbtSZj9wP/CF/JnZuU4rcpk7koLFv+X3/cCrwKw66+2CPP2FwmNDXq7fAAuANwKfy/N9LZf5r/n57kr9c3mfzenLgaML6cX6Lyik75vL2Fz/nD4X+F1e1w8CZ7W7LXbCowN+D+8plPWHwvezdKjvufCZCYV22ay2U3ceHVTHlrfvtjfeUTb4SyorlLQxurjOj2Jlfp6UX1ca8D3A4bkh3wwcO1S5pI3wzTn/YaSNayUA/B64s2reAbw9v+4D/hvwBLBfbiyvkzb0xY3rHcBlwJXAj4BrCz/aV3MD3Jf0B5P7c0N5EfhT0gZ+HXBZ/sztwHP59WXAx4GdgJeB7+WGuboyjxrr7gLgu4X32+X6B6nP/yHggML0XYD/SwoEk0kBqFL/A3L+Sv2fyOVVytwv139zmcB1wOxi/Qs/kEvb3f467dEBv4e7C/PZUDXfut9zIc8nCm23WW2n5jw6rI4tb9/d2gU0C1iSXy8BPlQjz9HAsohYFxHPk/ZgjpG0J7BLRPwi0lq/qvD5euXOAq6K5C7S3vfkBusapIZRuRzAKlLD+GJE7JofE4H/KMz7ceBISQLeB/xHRLwSEU+SNv7nkgLDJuCkiHgVeBJ4R/7Mu3N5m5cjIn5P6i7aCJxGaniVeQznUNJeDKSNfOVyBmkBI14kDSjYPi/r9YWyZ5G6mir178/lbb5EQq7/NcCs/Jkjchmb699AHcus3b+HXXM5tdT8noeof7PaTr15dFIdW65bA8DUiFgDkJ93r5Gn1l/wp+XHQI30ocqtLmsTqb8fYAfgzyTdJanWF/kMsH/h8z8h7YEdIWmH/DiEtKdQyROkkTW7AW8hHVaT870C/HEuY2NlWn79pvyZ9cBukt5ctXyvkc4jnED6oVfmMZxa63KWpEMk7ShpJ+CgXN7ySEMfK2UP9T3USt8NeCGXUUyv+HA++X29pOIfrMqs3b+H4md2knRf4fcwVN5BdWti26k3j+GWp5V1hBa37469H4CknwJ71Jh0XqNF1EiLIdJHUxbAfOBE4HTgNkmPVOX7GvBD0gb55Yj4pKSLgXNI3TBvIB0e1grG1fXaBfjPpCtLvkY6N/ClGnV9DbiadJi/PVv+efofpHMUD0TEqrwDNNyyV8qsVbd/JAWjTaRD4P8aERuq8tT7bL3lHWpd/xi4OiJekXQ2ae/piAbq3/W66PfwxxGxWtJ+wG0Mbp+1yh9p3UbTdkY7rZV1bHn77tgAEBF1/2gl6RlJe0bEmnzoubZGtgGgt/B+b1J//EB+XUyvbCDrlVv9d/6ngH/O9fwq8NVcrz7g3RGx+UuOiF9I+jBwQUR8MidvD3wzIr5YWKZbgH0iYq6k7YGzSf369+QHEXGrpHtJffTkMisb3Jvz87OkEULzIuJjkg4v5B8g7YF8J8/jzXkeW4mIC6qSBnLdlOt6EukEX3X91+bXxbKHuhRCrfRnSV0K2+e9pM35I6I44ulbwOBLzY1THf57KH5HleeV+fewS728VXXbBxhoVtsZYh5DLk8r69iW9t3KEw5j9QD+D1ufnLqkRp7JpH7xSfnxJDA5T7uXdPKqctLruKHKBY5n65Ne9+T0ScCE/HoK6az+ATXqsj1pb3xftpwQOrAqzzlsfZLpuvz6QLY+ybSSdIKpbpnAD9j6JNMn8uuLSV1IbyrOo4H13Sn137Mwv78E7mp3W+yER4f/Hg7qkLZTcx5lb99tb7yjbPC7AbfmBnZroSHPAL5dyPcx0kmZfuCMQvoM4FHSydhL2TLsrV65It3U4wnScLkZOf29+f1D+fnMIep8HPDrXMZ5Oe1C4IT8eqfcMPpJe/z7FT57Xv7ccvIIjXrev3o3AAAKdUlEQVRl5vT9chn9ucwJpMPJ9YX0rebRwDpva/1z+heBx/L6vh3Yv91tsRMenf576JC2U3ceHVTHlrfvyhdtZmYl062jgMzMbBt17ElggClTpkRPT09Tyt64cSM779w9t27rtvpC59T5/vvvfzYi3tLueph1mo4OAD09Pdx3331NKbuvr4/e3t6mlN0M3VZf6Jw6S/pNu+tg1oncBWRmVlIdfQRQNj0Lbqw7bf5Bm5hbY/qqhcc3s0pm1qHy/3EuIF2f69CIGHF3ybBHAJL2kXS7pMclPSbpr3P65HxN+xX5eVJOl6RFkvrzX5rfUyhrTs6/QtKckVbWzKyMlO7yd2VV8qPAX5EuJDkqjXQBbQLmR8Q7SX/6OEfSAaQ/htwaEdNJY4QX5PzHAtPzYx7wjbwAk4HzSdfOPxQ4vxI0zMxsZCLi8YhYvi1lDBsAImJNRDyQX79EulLlNEZ+pcCaVyPclsqbmdnojegcgKQe0qWG76bqSoGShrtSYCNX3EPSPNKRA1OnTqWvr28kVWzYhg0bmlb2aM0/aFPdaVPfWHt6py1DUSeuY7NuIulu0j/5JwKTJT2YJ30mIm7Z1vIbDgCSJpKuavmpiHhxiMvIb9NVByPdUHkxwIwZM6JZwwg7ZYhiUa2TvBXzD9rElx8Z/HWtOrW3iTXaNp24js26SUTMhHQOAJgbEXPHsvyGhoHmG4f/EPheRPwoJz9TuQlEg1cKbOSKe2Zm1iKNjAIScDnweET8XWHSUtJdqcjPNxTST8+jgQ4D1ueuoluAoyRNyid/j8ppZmY2QpL+UtIA6XaeN+ZLso9II11A7yPdQvCRQv/T54CFwHWSziRdH/+kPO0m0lXw+kn3oD0DICLWSfo86dKzABdGxKBr0ZuZ2dYioo90/4Zi2j8B/7Qt5Q4bACLiTmr33wMcWSN/kK6dXausK4ArRlJBMzNrDl8KwsyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSqqRm8JfIWmtpEcLaZMlLZO0Ij9PyumStEhSv6SHJb2n8Jk5Of8KSXNqzcvMzFqnkSOAK4FjqtIWALdGxHTg1vwe4Fhgen7MA74BKWAA5wMzgUOB8ytBw8zM2mPYABARdwDrqpJnAUvy6yXAhwrpV0VyF7CrpD2Bo4FlEbEuIp4HljE4qJiZWQttP8rPTY2INQARsUbS7jl9GvB0Id9ATquXPoikeaSjB6ZOnUpfX98oqzi0DRs2NK3s0Zp/0Ka606a+sfb0TluGok5cx2a2xWgDQD2qkRZDpA9OjFgMLAaYMWNG9Pb2jlnlivr6+mhW2aM1d8GNdafNP2gTX35k8Ne16tTeJtZo23TiOjazLUY7CuiZ3LVDfl6b0weAfQr59gZWD5FuZmZtMtoAsBSojOSZA9xQSD89jwY6DFifu4puAY6SNCmf/D0qp5mZWZsM2wUk6WqgF5giaYA0mmchcJ2kM4GngJNy9puA44B+4GXgDICIWCfp88C9Od+FEVF9YtnMzFpo2AAQEafUmXRkjbwBnFOnnCuAK0ZUOzMzaxr/E9jMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzEpqrG8IYy3WM8RNZOpZtfD4JtTEzLqNA0CTjGbDbGbWSu4CMjMrKQcAM7OScgAwMyspBwAzs5Jq+UlgSccAXwO2A74dEQtbXYeRGI8nc0e7TB49ZDa+tPQIQNJ2wNeBY4EDgFMkHdDKOpiZWdLqI4BDgf6IWAkg6RpgFvDLVsy8uOc7/6BNzB2He/fNNNIjh/kHbaK3OVUxszGgiGjdzKQTgWMi4qz8/jRgZkScW8gzD5iX3/4JsLxJ1ZkCPNukspuh2+oLnVPnt0bEW9pdCbNO0+ojANVI2yoCRcRiYHHTKyLdFxEzmj2fsdJt9YXurLNZmbR6FNAAsE/h/d7A6hbXwczMaH0AuBeYLmlfSTsCs4GlLa6DmZnR4i6giNgk6VzgFtIw0Csi4rFW1qGg6d1MY6zb6gvdWWez0mjpSWAzM+sc/iewmVlJOQCYmZVUVwYASZMlLZO0Ij9PqpNvTs6zQtKcQvrBkh6R1C9pkSQNVa6SRTn/w5LeUyjrD5IezI+lVfM/RtLy/LkFNeo3QdK1efrdknoK0z6b05dLOnq4MvOJ9btz3a/NJ9mHnEedddYJdZ4r6XeF9XrWUHU2s1GKiK57AJcAC/LrBcDFNfJMBlbm50n59aQ87R7gcNL/Em4Gjh2qXOC4nE/AYcDdhflsqFPH7YAngP2AHYGHgAOq8nwCuCy/ng1cm18fkPNPAPbN5Ww3VJnAdcDs/Poy4ONDzaPD6zwXuLTd7cwPP8b7oyuPAEiXj1iSXy8BPlQjz9HAsohYFxHPA8uAYyTtCewSEb+IiACuKny+XrmzgKsiuQvYNZczlM2XvYiIV4HKZS/qLcf1wJH5aGQWcE1EvBIRTwL9ubyaZebPHJHLqFX3WvPo5DqbWQt0awCYGhFrAPLz7jXyTAOeLrwfyGnT8uvq9KHKrVcWwE6S7pN0l6TiBmyozwzKExGbgPXAbsPUvVb6bsALuYzqedWbRy2dUmeAD+futuslFf88aGZjpGPvCSzpp8AeNSad12gRNdJiiPTRlAXwxxGxWtJ+wG2SHomIJxqcz0jrWCtgD7dMI1neTqnzj4GrI+IVSWeTjg6OqFljMxu1jg0AEfGBetMkPSNpz4hYk7ti1tbINgBbXYxyb6Avp+9dlV65HEW9cutewiIiKs8rJfUB7yb1eTdy2YtKngFJ2wNvBtYN89la6c+SuqW2z3vUxfz15lFLR9Q5Ip4r5P8WcHGd+prZNujWLqClQGVUzxzghhp5bgGOkjQpj+Y5Crgld+28JOmw3A99euHz9cpdCpyeRwMdBqzPQWKSpAkAkqYA72PLpa0buexFcX4nArfl8xJLgdl5xM2+wHTSieuaZebP3J7LqFX3WvOopSPqXHV+5QTg8Tr1NbNt0e6z0KN5kPqPbwVW5OfJOX0G6S5jlXwfI52M7AfOKKTPAB4l7alfypZ/RNcrV6Qb2TwBPALMyOnvze8fys9nVtXzOODX+XPn5bQLgRPy652AH+T63QPsV/jseflzy8mjlOqVmdP3y2X05zInDDePOuu2E+r8ReCxvF5vB/Zvd5vzw4/x+PClIMzMSqpbu4DMzGwbOQCYmZWUA4CZWUk5AJiZlZQDgI0Lkk6S9Jik1yX5PsRmDXAAsK4jqVfSlVXJjwJ/BdzR+hqZdaeO/Sew2UhExOMA9a9zZ2bVfARgZlZSPgKwriHpbtL9BiYCkyU9mCd9JiJuaV/NzLqTA4B1jYiYCekcADA3Iua2tUJmXc5dQGZmJeUAYOOCpL+UNEC61eeNktwlZDYMXwzOzKykfARgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZS/x+LBUEIJjb9RwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_bg_intensity = df_g_intensity.copy()\n",
    "for column in df_bg_intensity.columns[1:]:\n",
    "    df_bg_intensity[column] = df_b_intensity.loc[:,[column+str(e+1) for e in range(8)]].sum(axis=1)\n",
    "(df_bg_intensity / df_g_intensity).hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get  normalized iBAQs for 4 groups\n",
    "20190921"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### notes\n",
    "\n",
    "```\n",
    "Segment 3, all samples\n",
    "1.  All of the samples had a predominant protein band in them (see attached images).  \n",
    "1a.  This band was used to define the cut borders for \"gel segment 3;\"  in our lane excisions, \"segment 3\" was cut just above and just below this band, so that the segment contained only the band.\n",
    "1b.  This band (now known as segment 3) was so abundant that is was necessary to dilute it 1:20 prior to running it on the LC column.\n",
    "1c.  However, the other lane fractions (segments #1, #2, #4, #5, #6, #7,and #8) did not require this dilution, and diluting them would have greatly diminished your sensitivity.  \n",
    "1d.  Thus, they were injected without further dilution.\n",
    "1e.  Summarizing, segment 3 for each sample was diluted 1:20 prior to injection.\n",
    " \n",
    "2.  For LFQ analysis, this doesn't require any correction;  LFQ strives for relative expression between samples.  \n",
    "\n",
    "3.  For iBAQ, however, the iBAQ intensities should probably multiplied 10X.  This is an imperfect solution (due to the finer points of electrospray physics) but it's a solution.\n",
    "Adult samples\n",
    "1. The adult samples have a similar problem, in that there are 3 other gel regions / bands that were too abundant to load without dilution:  segments #1, #2, #7, and #8.  Thus, they (and segment #3) were diluted 1:20 before injection\n",
    "2.  These could be handled as proposed above for iBAQ. \n",
    "\n",
    "3.  Alternatively, you could do your quantitative thinking based on the band images, that is, the gels clearly show that the adults express proteins that are not very abundant in larvae, and MS was used to ID these proteins.\n",
    "\n",
    "4.  It's not clear to me how you might best perform LFQ analyses of adult segments.  \n",
    "4a.  Including adult segments 1,2,7,8 in an LFQ normalization would be awkward with regards the thinking, the math, and the explanation;  it's hard for me to decide if the LFQ values can fairly be multiplied by 10 here, then compared to the other lanes--maybe OK?\n",
    "4b.  Alternativley, you could leave 1,2,7,8 from all stages out of the LFQanalysis, performing a separate LFQ analysis of just the larvae stages, all segments?\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check best factor for diluted bands\n",
    "\n",
    "for three replicates in Adult, theoretically, we should multiply iBAQs of bands 1,2,7,8 by 20 and then sum iBAQs in one lane to get the total iBAQ per lane. However, dilution of 20x, the iBAQ values may be not 1/20 than expected. It could be greater than 1/20. Dr. Hartson suggest to multiply by 10. Check values from 10 to 20, and then the correlation of the three samples, and see which factor makes the correlation value largest.\n",
    "\n",
    "```python\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "file_band = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20190921MsDevBand_ori.xlsx\"\n",
    "\n",
    "dfDevBand = pd.read_excel(file_band, index_col=0)\n",
    "\n",
    "samples = ['AdultA', 'AdultB','AdultC']\n",
    "\n",
    "factors = [1 + e * 0.5 for e in range(40)] #for column 1,2,3,7,8\n",
    "\n",
    "def calculateCor(factor):\n",
    "    tdf = pd.DataFrame()\n",
    "    for sample in samples:\n",
    "        tdf[sample] = dfDevBand[['iBAQ '+sample+i for i in '1378']].sum(axis=1) * factor + dfDevBand[['iBAQ '+sample+i for i in '2456']].sum(axis=1)\n",
    "        tdf[sample] = tdf[sample] / tdf[sample].sum() * 1e6\n",
    "#    tdf = tdf[tdf.max(axis=1) > 1]\n",
    "    cor = tdf.corr(method='spearman')\n",
    "    return cor.values.mean()\n",
    "\n",
    "cors = [calculateCor(i) for i in factors]\n",
    "\n",
    "plt.plot(factors, cors)\n",
    "\n",
    "keys = ['Bar', 'PupaD4','PupaD13']\n",
    "def calculateCorKey(factor):\n",
    "    corrs = []\n",
    "    for k in keys:\n",
    "        tdf = pd.DataFrame()\n",
    "        for s in 'ABC':\n",
    "            sample = k+s\n",
    "            tdf[sample] = dfDevBand[['iBAQ '+sample+i for i in '3']].sum(axis=1) * factor + dfDevBand[['iBAQ '+sample+i for i in '1245678']].sum(axis=1)\n",
    "            tdf[sample] = tdf[sample] / tdf[sample].sum() * 1e6\n",
    "        corrs.append(tdf.corr(method='spearman').values.mean())\n",
    "    return np.mean(corrs)\n",
    "\n",
    "cors2 = [calculateCorKey(i) for i in factors]\n",
    "plt.plot(factors, cors2)\n",
    "\n",
    "```\n",
    "\n",
    "conclusion:\n",
    "\n",
    "I checked to use pearson correlation and spearman correlation. The result is not very consistent. Hard to tell. Go on use the original dilution fold.\n",
    "\n",
    "After check Steve's notes, for sample Bar and P4, cut to 8 pieces. segment 3 diluted 20x. For sample P13 and Adult, cut to 9 pieces, with two pieces pooled together to make band 4. band 1, 3, 7 and 8 were diluted 15x."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### membrane proteins\n",
    "Combine two sheet, group and band to one big file.\n",
    "\n",
    "bands were not diluted. calculate ppm (parts per million) value for proteins in each band and each lane. First calculate total iBAQ for each lane, then iBAQ/(total iBAQ) * 1e6 to get ppm value.\n",
    "\n",
    "add column \"Protein\". For elements in column \"Protein IDs\", keep the one with the highest peptide count (in column \"Peptide counts (all)\"). If there is both MCOT and OGS IDs, use MCOT id if the count of peptide of MCOT - count of OGS >= 5.\n",
    "\n",
    "add column \"Annotated\" and \"description\" based on column \"Protein\", to see if the protein is already annotated in previous work and the protein name in NCBI.\n",
    "\n",
    "add column \"protein_seq\", include protein sequences.\n",
    "\n",
    "add column \"protein_MW\", include protein molecular weight.\n",
    "\n",
    "add column \"GO\", include gene ontology terms for gene based on column \"Protein\".\n",
    "\n",
    "add column \"KO\", include KEGG KO terms for gene based on column \"Protein\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import SeqUtils\n",
    "\n",
    "\n",
    "#file of each band and each lane\n",
    "fMemBand = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\membrane\\20190921MsMemBand_ori.xlsx\"\n",
    "fMemGroup = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\membrane\\20190921MsMemGroup_ori.xlsx\"\n",
    "#read in file\n",
    "dfMemBand = pd.read_excel(fMemBand, index_col=0)\n",
    "dfMemGroup = pd.read_excel(fMemGroup, index_col=0)\n",
    "df = dfMemBand.copy()\n",
    "#4 samples keywords. FP, fat body precipitates; FS, fat body supernatants; HP, hemocyte precipitates; HS, hemocyte supernatants\n",
    "samples = 'FP FS HP HS'.split()\n",
    "# combine columns in df_lane to df\n",
    "columns_band = dfMemBand.columns\n",
    "for c in dfMemGroup.columns:\n",
    "    if c not in columns_band:\n",
    "        df[c] = dfMemGroup[c]\n",
    "\n",
    "# add a column 'Protein', only one protein ID kept in the 'Protein IDs' column. Keep the one have the largest 'Peptide counts (all)'. If more than one proteins have the largest 'Peptide counts (all)' value, keep OGS2 ID\n",
    "def getProteinID(r, adjustForOGS = 0.5):\n",
    "    '''\n",
    "    r is one row of dataframe\n",
    "    proteinIDs looks like 'Msex2.00001-RA;MCOT.C00001.1.0.TOO5O;MCOT.C00001.3.0.OOO5O;MCOT.C00001.2.0.TOO5O'\n",
    "    keep Msex2 id if there is Msex2 id, otherwise, keep the first ID\n",
    "    sort by counts. if OGS, counts + adjustForOGS to perfer OGS\n",
    "    '''\n",
    "    tdf = pd.DataFrame()\n",
    "    proteins = r['Protein IDs'].split(';')\n",
    "    counts = r['Peptide counts (all)'].split(';')\n",
    "    counts =[float(e) for e in counts]\n",
    "    tdf['protein'] = proteins\n",
    "    tdf['counts'] = counts\n",
    "    tdf['counts_adjust'] = tdf.apply(lambda x:x['counts'] + adjustForOGS if x['protein'].startswith('Msex2.') else x['counts'], axis=1)\n",
    "    tdf = tdf.sort_values(by='counts_adjust', ascending=False)\n",
    "    protein = tdf.iloc[0]['protein']\n",
    "    return protein.split(':')[0]\n",
    "\n",
    "df['Protein'] = df.apply(getProteinID, axis='columns')\n",
    "# add a column \"MCOTonly\" to see if MCOT is the only protein ID\n",
    "df['MCOTonly'] = df['Protein IDs'].apply(lambda x:'Msex2.' not in x)\n",
    "\n",
    "# add protein names\n",
    "file_annotated = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsAnnotatedSeq.xlsx\"\n",
    "file_names = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsMCOT_OGS2.xlsx\"\n",
    "df_annotated = pd.read_excel(file_annotated,sheet_name = '2Anno')\n",
    "dc_name2anno = dict(zip(df_annotated['seq2_id'], df_annotated['seq1_id']))\n",
    "df_names = pd.read_excel(file_names, sheet_name = 'Sheet1')\n",
    "dc_name2description = dict(zip(df_names['Query id'], df_names['description']))\n",
    "def addAnnoDes(df):\n",
    "    '''\n",
    "    add two column, 'Annotated' and 'description' for input dataframe\n",
    "    '''\n",
    "    df['Annotated'] = df['Protein'].apply(lambda x: dc_name2anno[x] if x in dc_name2anno else '')\n",
    "    df['description'] = df['Protein'].apply(lambda x: dc_name2description[x] if x in dc_name2description else '')\n",
    "    return df\n",
    "\n",
    "df = addAnnoDes(df)\n",
    "\n",
    "# add GO terms\n",
    "file_GO = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsMCOTOGS2_GO.csv\"\n",
    "df_GO = pd.read_csv(file_GO, sep='\\t',header=None)\n",
    "dc_GO = dict(zip(df_GO[0], df_GO[1]))\n",
    "df['GO'] = df['Protein'].apply(lambda x: dc_GO[x] if x in dc_GO else '')\n",
    "\n",
    "# add KO terms\n",
    "file_KO = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190922MsMCOT_OGS_KEGG_KO.txt\"\n",
    "df_KO = pd.read_csv(file_KO, sep='\\t',header=None)\n",
    "dc_KO = {}\n",
    "for k,v in df_KO.groupby(0):\n",
    "    dc_KO[k] = '|'.join(list(v[1]))\n",
    "\n",
    "df['KO'] = df['Protein'].apply(lambda x: dc_KO[x] if x in dc_KO else '')\n",
    "\n",
    "# add protein sequence and \"MolecularWeight\"\n",
    "dc_seqs = {}\n",
    "file_ogs = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\ms_ogs_proteins.fa\"\n",
    "file_mcot = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta\"\n",
    "for f in [file_ogs, file_mcot]:\n",
    "    for s in SeqIO.parse(f,'fasta'):\n",
    "        seq = str(s.seq)\n",
    "        if seq[-1] == '*':\n",
    "            seq = seq[:-1]\n",
    "        dc_seqs[s.id] = seq\n",
    "df['seq'] = df['Protein'].apply(lambda x:dc_seqs[x])\n",
    "df['MolecularWeight'] = df['seq'].apply(lambda x: SeqUtils.molecular_weight(x.replace('X',''),seq_type='protein'))\n",
    "\n",
    "# add columns of if with signalP\n",
    "file_sigP = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190303ms_signalP.tab\"\n",
    "file_sig3L = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190303ms_signal3L.tab\"\n",
    "df_sig = pd.concat([pd.read_csv(f, sep='\\t') for f in [file_sig3L, file_sigP]])\n",
    "dc_sig = dict(zip(df_sig['ms_id'], df_sig['Start_nonSignalAA']))\n",
    "df['Start_nonSignalAA'] = df['Protein'].apply(lambda x:dc_sig[x] if x in dc_sig else '')\n",
    "df['SignalP'] = df['Start_nonSignalAA'] != ''\n",
    "\n",
    "# add column immuneRelated\n",
    "file_immune = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsImmuneRelatedSeq_MCOTOGS.csv\"\n",
    "df_immune = pd.read_csv(file_immune, sep='\\t')\n",
    "dc_immune = dict(zip(df_immune['seq2_id'], df_immune['seq1_id']))\n",
    "df['immuneRelated'] = df['Protein'].apply(lambda x:x in dc_immune)\n",
    "\n",
    "df.to_excel(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\membrane\\20191016MsMem_all.xlsx\")\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hemolymph developmental, combine groups and bands excel\n",
    "\n",
    "I did not take dilution into consideration for now. That is for comparing protein of different stages.\n",
    "\n",
    "```python\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import SeqUtils\n",
    "\n",
    "\n",
    "#file of each band and each lane\n",
    "fMemBand = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20190921MsDevBand_ori.xlsx\"\n",
    "fMemGroup = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20190921MsDevGroup_ori.xlsx\"\n",
    "#read in file\n",
    "dfMemBand = pd.read_excel(fMemBand, index_col=0)\n",
    "dfMemGroup = pd.read_excel(fMemGroup, index_col=0)\n",
    "df = dfMemBand.copy()\n",
    "#4 samples keywords. FP, fat body precipitates; FS, fat body supernatants; HP, hemocyte precipitates; HS, hemocyte supernatants\n",
    "samples = 'FP FS HP HS'.split()\n",
    "# combine columns in df_lane to df\n",
    "columns_band = dfMemBand.columns\n",
    "for c in dfMemGroup.columns:\n",
    "    if c not in columns_band:\n",
    "        df[c] = dfMemGroup[c]\n",
    "\n",
    "# dc_sig to save signal peptide info\n",
    "file_sigP = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190303ms_signalP.tab\"\n",
    "file_sig3L = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190303ms_signal3L.tab\"\n",
    "df_sig = pd.concat([pd.read_csv(f, sep='\\t') for f in [file_sig3L, file_sigP]])\n",
    "dc_sig = dict(zip(df_sig['ms_id'], df_sig['Start_nonSignalAA']))\n",
    "\n",
    "# add a column 'Protein', only one protein ID kept in the 'Protein IDs' column. Keep the one have the largest 'Peptide counts (all)'. If more than one proteins have the largest 'Peptide counts (all)' value, keep OGS2 ID\n",
    "def getProteinID(r, adjustForOGS = 2.5):\n",
    "    '''\n",
    "    r is one row of dataframe\n",
    "    proteinIDs looks like 'Msex2.00001-RA;MCOT.C00001.1.0.TOO5O;MCOT.C00001.3.0.OOO5O;MCOT.C00001.2.0.TOO5O'\n",
    "    add a column if this signal peptide. Prefer those with signal peptide.\n",
    "    first keep proteinIDs with highest peptide count\n",
    "    sort by signal peptide, counts and if it is OGS2\n",
    "    '''\n",
    "    tdf = pd.DataFrame()\n",
    "    proteins = r['Protein IDs'].split(';')\n",
    "    counts = r['Peptide counts (all)'].split(';')\n",
    "    counts =[float(e) for e in counts]\n",
    "    tdf['protein'] = proteins\n",
    "    tdf['counts'] = counts\n",
    "    counts_max = tdf['counts'].max()\n",
    "    tdf = tdf[tdf['counts'] >= counts_max]\n",
    "    tdf['isOGS2'] = tdf.apply(lambda x:1 if x['protein'].startswith('Msex2.') else 0, axis=1)\n",
    "    tdf['signalP'] = tdf['protein'].apply(lambda x:1 if x in dc_sig else 0)\n",
    "    tdf = tdf.sort_values(by=['signalP','isOGS2'], ascending=False)\n",
    "    protein = tdf.iloc[0]['protein']\n",
    "    return protein.split(':')[0]\n",
    "\n",
    "df['Protein'] = df.apply(getProteinID, axis='columns')\n",
    "# add a column \"MCOTonly\" to see if MCOT is the only protein ID\n",
    "df['MCOTonly'] = df['Protein IDs'].apply(lambda x:'Msex2.' not in x)\n",
    "\n",
    "# add protein names\n",
    "file_annotated = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsAnnotatedSeq.xlsx\"\n",
    "file_names = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20181129MsMCOT_OGS2.xlsx\"\n",
    "df_annotated = pd.read_excel(file_annotated,sheet_name = '2Anno')\n",
    "dc_name2anno = dict(zip(df_annotated['seq2_id'], df_annotated['seq1_id']))\n",
    "df_names = pd.read_excel(file_names, sheet_name = 'Sheet1')\n",
    "dc_name2description = dict(zip(df_names['Query id'], df_names['description']))\n",
    "def addAnnoDes(df):\n",
    "    '''\n",
    "    add two column, 'Annotated' and 'description' for input dataframe\n",
    "    '''\n",
    "    df['Annotated'] = df['Protein'].apply(lambda x: dc_name2anno[x] if x in dc_name2anno else '')\n",
    "    df['description'] = df['Protein'].apply(lambda x: dc_name2description[x] if x in dc_name2description else '')\n",
    "    return df\n",
    "\n",
    "df = addAnnoDes(df)\n",
    "\n",
    "# add GO terms\n",
    "file_GO = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsMCOTOGS2_GO.csv\"\n",
    "df_GO = pd.read_csv(file_GO, sep='\\t',header=None)\n",
    "dc_GO = dict(zip(df_GO[0], df_GO[1]))\n",
    "df['GO'] = df['Protein'].apply(lambda x: dc_GO[x] if x in dc_GO else '')\n",
    "\n",
    "# add KO terms\n",
    "file_KO = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190922MsMCOT_OGS_KEGG_KO.txt\"\n",
    "df_KO = pd.read_csv(file_KO, sep='\\t',header=None)\n",
    "dc_KO = {}\n",
    "for k,v in df_KO.groupby(0):\n",
    "    dc_KO[k] = '|'.join(list(v[1]))\n",
    "\n",
    "df['KO'] = df['Protein'].apply(lambda x: dc_KO[x] if x in dc_KO else '')\n",
    "\n",
    "# add protein sequence and \"MolecularWeight\"\n",
    "dc_seqs = {}\n",
    "file_ogs = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\ms_ogs_proteins.fa\"\n",
    "file_mcot = r\"E:\\Insects\\ManducaSexta\\MCOT_OGS\\20150105MCOT30303proteinSequenceNewIDfastaRemoveOtherSpecies.fasta\"\n",
    "for f in [file_ogs, file_mcot]:\n",
    "    for s in SeqIO.parse(f,'fasta'):\n",
    "        seq = str(s.seq)\n",
    "        if seq[-1] == '*':\n",
    "            seq = seq[:-1]\n",
    "        dc_seqs[s.id] = seq\n",
    "df['seq'] = df['Protein'].apply(lambda x:dc_seqs[x])\n",
    "df['MolecularWeight'] = df['seq'].apply(lambda x: SeqUtils.molecular_weight(x.replace('X',''),seq_type='protein'))\n",
    "\n",
    "# add columns of if with signalP\n",
    "df['Start_nonSignalAA'] = df['Protein'].apply(lambda x:dc_sig[x] if x in dc_sig else '')\n",
    "df['SignalP'] = df['Start_nonSignalAA'] != ''\n",
    "\n",
    "# add column immuneRelated\n",
    "file_immune = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20181128ManducaGeneNames\\20190921MsImmuneRelatedSeq_MCOTOGS.csv\"\n",
    "df_immune = pd.read_csv(file_immune, sep='\\t')\n",
    "dc_immune = dict(zip(df_immune['seq2_id'], df_immune['seq1_id']))\n",
    "df['immuneRelated'] = df['Protein'].apply(lambda x:x in dc_immune)\n",
    "\n",
    "df.to_excel(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all_1.xlsx\") #save content to main file manually\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### update the `Annotated` column\n",
    "20191130, 53 SPSPHs were added to the annotated sequences. Update the `Annotated` column. Note, the `20181129MsAnnotatedSeq.xlsx` is already updated. Just need to run the code above and copy the \"Annotated\" column to the old file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calculate normalized ppm for proteins for Hemolymph development"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "below are dilution notes\n",
    "```\n",
    "Bar\n",
    "gel fraction 1..... was run without dilution\n",
    "gel fraction 2 ..... was run without dilution \n",
    "gel fraction 3 ..... was diluted further 1:20 (2 ul + 38 ul)\n",
    "gel fraction 4 ..... was run without dilution \n",
    "gel fraction 5 ..... was run without dilution \n",
    "gel fraction 6 ..... was run without dilution \n",
    "gel fraction 7 ..... was run without dilution \n",
    "gel fraction 8 ..... was run without dilution \n",
    " \n",
    "P4\n",
    "gel fraction 1..... was run without dilution\n",
    "gel fraction 2 ..... was run without dilution \n",
    "gel fraction 3 ..... was diluted further 1:20 (2 ul + 38 ul) \n",
    "gel fraction 4 ..... was run without dilution \n",
    "gel fraction 5 ..... was run without dilution \n",
    "gel fraction 6 ..... was run without dilution \n",
    "gel fraction 7 ..... was run without dilution \n",
    "gel fraction 8 ..... was run without dilution \n",
    " \n",
    "P13\n",
    "gel fraction 1..... was run without dilution\n",
    "gel fraction 2 ..... was run without dilution \n",
    "gel fraction 3 ..... was diluted further 1:20 (2 ul + 38 ul) \n",
    "gel fraction 4 ..... was run without dilution \n",
    "gel fraction 5 ..... was run without dilution \n",
    "gel fraction 6 ..... was run without dilution \n",
    "gel fraction 7 ..... was run without dilution \n",
    "gel fraction 8 ..... was run without dilution \n",
    " \n",
    "Adult\n",
    "gel fraction 1..... was diluted further 1:20 (2 ul + 38 ul) \n",
    "gel fraction 2 ..... was run without dilution \n",
    "gel fraction 3 ..... was diluted further 1:20 (2 ul + 38 ul) \n",
    "gel fraction 4 ..... was run without dilution \n",
    "gel fraction 5 ..... was run without dilution \n",
    "gel fraction 6 ..... was run without dilution\n",
    "gel fraction 7 ..... was diluted further 1:20 (2 ul + 38 ul) \n",
    "gel fraction 8 ..... was diluted further 1:20 (2 ul + 38 ul)  \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "total intensity for each lane\n",
    "\n",
    "```\n",
    "Bar-A        2.730558e+11\n",
    "Bar-B        2.484587e+11\n",
    "Bar-C        2.980367e+11\n",
    "PupaD4-A     2.210165e+11\n",
    "PupaD4-B     1.894565e+11\n",
    "PupaD4-C     1.731804e+11\n",
    "PupaD13-A    2.216864e+11\n",
    "PupaD13-B    1.937883e+11\n",
    "PupaD13-C    2.781827e+11\n",
    "Adult-A      2.363084e+11\n",
    "Adult-B      2.011981e+11\n",
    "Adult-C      2.275128e+11\n",
    "```\n",
    "very conserved. calculate ppm for each protein. similar to TPM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# 32 groups\n",
    "lanes = []\n",
    "for stage in stages:\n",
    "    for rep in 'ABC':\n",
    "        lanes.append(f'{stage}-{rep}')\n",
    "\n",
    "# calculate protein iBAQ in each lane\n",
    "df_iBAQ = df[['Protein']].copy()\n",
    "for lane in lanes:\n",
    "    stage, rep = lane.split('-')\n",
    "    if stage != 'Adult':\n",
    "        columns_normal = [1,2,4,5,6,7,8]\n",
    "        columns_diluted = [3]\n",
    "    else:\n",
    "        columns_normal = [2,4,5,6]\n",
    "        columns_diluted = [1,3,7,8]\n",
    "    df_iBAQ[lane] = 0\n",
    "    for band in columns_normal:\n",
    "        df_iBAQ[lane] = df_iBAQ[lane] + df[f'iBAQ {stage}{rep}{band}']\n",
    "    for band in columns_diluted:\n",
    "        df_iBAQ[lane] = df_iBAQ[lane] + 20 * df[f'iBAQ {stage}{rep}{band}']\n",
    "\n",
    "# for Adult, the loading volumn is 1.5, instead of 1 for others. control that\n",
    "for rep in 'ABC':\n",
    "    df_iBAQ[f'Adult-{rep}'] = df_iBAQ[f'Adult-{rep}'] / 1.5\n",
    "\n",
    "# print total iBAQs for each lane\n",
    "print(df_iBAQ.iloc[:,1:].sum())\n",
    "\n",
    "# calculate PPM for each protein, similar to TPM\n",
    "for lane in lanes:\n",
    "    df_iBAQ['PPM '+lane] = df_iBAQ[lane] / df_iBAQ[lane].sum() * 1e6\n",
    "\n",
    "# save result to csv\n",
    "df_iBAQ.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191104Protein_iBAQ_PPM.csv\", sep='\\t') # add columns to 20191016MsDev_all.xlsx manually\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hemolymph Development\n",
    "\n",
    "get some statistics to make figures."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### some basic counts about proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### protein identified in Bar, P4, P13 and Adult\n",
    "\n",
    "venn diagram showing the number of identified proteins, identified secreted/cellular proteins, immune related proteins among 4 stages."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "outfile = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.added.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "\n",
    "# count the number with Signal Peptide\n",
    "print('total number of proteins with signal peptide:', df['SignalP'].sum())\n",
    "print('total number of proteins with no signal peptide:', (~df['SignalP']).sum())\n",
    "\n",
    "# count the number immune related\n",
    "print('total number of proteins immune-related:', df['immuneRelated'].sum())\n",
    "\n",
    "\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# count number of identified proteins in four stages\n",
    "df_stages = pd.DataFrame()\n",
    "for key in stages:\n",
    "    columns = [f'Unique peptides {key}{i}' for i in 'ABC']\n",
    "    df[f'Unique peptides {key}'] = df[columns].sum(axis=1)\n",
    "    df_stages.loc[key,'Proteins identified'] = (df[f'Unique peptides {key}'] > 0).sum()\n",
    "    df_stages.loc[key,'Proteins identified (secreated)'] = ((df[f'Unique peptides {key}'] > 0) & (df['SignalP'])).sum()\n",
    "    df_stages.loc[key,'Proteins identified (cellular)'] = ((df[f'Unique peptides {key}'] > 0) & (~df['SignalP'])).sum()\n",
    "    df_stages.loc[key,'Proteins identified (immune)'] = ((df[f'Unique peptides {key}'] > 0) & (df['immuneRelated'])).sum()\n",
    "\n",
    "# count grouping of proteins.\n",
    "columns = [f'Unique peptides {key}' for key in stages]\n",
    "def getIdentificationStr(r):\n",
    "    '''\n",
    "    for each row, generate a string with like \"+---\" to show if a protein is expressed in Bar, PupaD4, PupaD13 and Adult\n",
    "    '''\n",
    "    elements = [r[i] for i in columns]\n",
    "    elements = ['+' if i > 0 else '-' for i in elements]\n",
    "    return ''.join(elements)\n",
    "df['Proteins identified Changes'] = df.apply(getIdentificationStr, axis=1)\n",
    "print(df['Proteins identified Changes'].value_counts())\n",
    "\n",
    "df_stages = df_stages[['Proteins identified','Proteins identified (cellular)', 'Proteins identified (secreated)', 'Proteins identified (immune)']]\n",
    "\n",
    "df_stages.loc['Total', 'Proteins identified'] = df.shape[0]\n",
    "df_stages.loc['Total', 'Proteins identified (cellular)'] = df.shape[0] - df['SignalP'].sum()\n",
    "df_stages.loc['Total', 'Proteins identified (secreated)'] = df['SignalP'].sum()\n",
    "df_stages.loc['Total', 'Proteins identified (immune)'] = df['immuneRelated'].sum()\n",
    "df_stages.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027stages_protein_count.csv\", sep='\\t')\n",
    "\n",
    "#df.to_excel(outfile) #save file to take a look carefully\n",
    "## check iBAQ for different \"Proteins identified Changes\"\n",
    "columns = [f'iBAQ {key}' for key in stages]\n",
    "df_Changes = pd.DataFrame()\n",
    "for k, v in df.groupby('Proteins identified Changes'):\n",
    "    for key in stages:\n",
    "        columns = [f'iBAQ {key}{i}' for i in 'ABC']\n",
    "        avg = v[columns].mean(axis=1).mean()\n",
    "        df_Changes.loc[k,key] = avg\n",
    "print(df_Changes)\n",
    "## plot venn diagram\n",
    "from pyvenn import venn\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['pdf.fonttype'] =  42\n",
    "labels = venn.get_labels([list(df[df[f'Unique peptides {key}']>0].index) for key in stages])\n",
    "fig, ax = venn.venn4(labels, names=['Bar', 'P4', 'P13', 'Adult'])\n",
    "fig.show()\n",
    "fig.savefig(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Figures\\venn_diagram_detected_proteins_all.pdf\")\n",
    "plt.close()\n",
    "\n",
    "## plot venn diagram for secreted\n",
    "labels = venn.get_labels([list(df[(df[f'Unique peptides {key}']>0) & df['SignalP']].index) for key in stages])\n",
    "fig, ax = venn.venn4(labels, names=['Bar', 'P4', 'P13', 'Adult'])\n",
    "fig.show()\n",
    "fig.savefig(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Figures\\venn_diagram_detected_proteins_signalP.pdf\")\n",
    "plt.close()\n",
    "\n",
    "## plot venn diagram for cellular\n",
    "labels = venn.get_labels([list(df[(df[f'Unique peptides {key}']>0) & ~df['SignalP']].index) for key in stages])\n",
    "fig, ax = venn.venn4(labels, names=['Bar', 'P4', 'P13', 'Adult'])\n",
    "fig.show()\n",
    "fig.savefig(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Figures\\venn_diagram_detected_proteins_cellular.pdf\")\n",
    "plt.close()\n",
    "\n",
    "\n",
    "## plot venn diagram for immune-related\n",
    "labels = venn.get_labels([list(df[(df[f'Unique peptides {key}']>0) & df['immuneRelated']].index) for key in stages])\n",
    "fig, ax = venn.venn4(labels, names=['Bar', 'P4', 'P13', 'Adult'])\n",
    "fig.show()\n",
    "fig.savefig(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Figures\\venn_diagram_detected_proteins_immune.pdf\")\n",
    "plt.close()\n",
    "\n",
    "# number of identified proteins in each stage\n",
    "plt.bar(df_stages.index, df_stages['Proteins identified (secreated)'])\n",
    "plt.bar(df_stages.index, df_stages['Proteins identified (cellular)'], bottom=df_stages['Proteins identified (secreated)'])\n",
    "plt.savefig(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Figures\\20191027bar_secreted_cellular_4stages.pdf\")\n",
    "plt.close()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### number of identified proteins in each gel slice"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "8 * 4 = 32 biological groups. \n",
    "* Number of identified proteins in each group. all, cellular and secreted\n",
    "* Get a matrix of shared identified proteins among the 32 groups. all, cellular and secreted. Protein identified if with >=1 unique peptides as identified.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "import itertools\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "outfile = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.added.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# 32 groups\n",
    "groups = []\n",
    "for stage in stages:\n",
    "    for band in range(1,9):\n",
    "        groups.append(f'{stage}-{band}')\n",
    "\n",
    "# get number of identified proteins for each group\n",
    "group_columns = ['protein_ids', 'protein_count', 'protein_ids_cellular','protein_count_cellular','protein_ids_secreted', 'protein_count_secreted']\n",
    "df_group = pd.DataFrame(index=groups, columns=group_columns)\n",
    "for group in groups:\n",
    "    stage, band = group.split('-')\n",
    "    columns = [f'Unique peptides {stage}{rep}{band}' for rep in 'ABC']\n",
    "    expressed = df[columns].sum(axis=1)\n",
    "    df_group.loc[group,'protein_ids'] = set(df[expressed > 0].index)\n",
    "    df_group.loc[group, 'protein_count'] = len(df_group.loc[group,'protein_ids'])\n",
    "    df_group.loc[group, 'protein_ids_cellular'] = set(df[(expressed>0) & ~df['SignalP']].index)\n",
    "    df_group.loc[group, 'protein_count_cellular'] = len(df_group.loc[group, 'protein_ids_cellular'])\n",
    "    df_group.loc[group, 'protein_ids_secreted'] = set(df[(expressed>0) & df['SignalP']].index)\n",
    "    df_group.loc[group, 'protein_count_secreted'] = len(df_group.loc[group, 'protein_ids_secreted'])\n",
    "\n",
    "# save df_group, only the counts\n",
    "df_group.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_count.csv', sep='\\t',columns=['protein_count', 'protein_count_cellular','protein_count_secreted'])\n",
    "\n",
    "# get matrix of shared all proteins between gel_bands\n",
    "df_share_all = pd.DataFrame(index=groups, columns=groups)\n",
    "df_share_cellular = pd.DataFrame(index=groups, columns=groups)\n",
    "df_share_secreted = pd.DataFrame(index=groups, columns=groups)\n",
    "for k1, k2 in itertools.product(df_group.index,repeat=2):\n",
    "    df_share_all.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids'] & df_group.loc[k2, 'protein_ids'])\n",
    "    df_share_cellular.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids_cellular'] & df_group.loc[k2, 'protein_ids_cellular'])\n",
    "    df_share_secreted.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids_secreted'] & df_group.loc[k2, 'protein_ids_secreted'])\n",
    "\n",
    "df_share_all.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_share_all.csv', sep='\\t')\n",
    "df_share_cellular.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_share_cellular.csv', sep='\\t')\n",
    "df_share_secreted.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_share_secreted.csv', sep='\\t')\n",
    "\n",
    "# matrix of shared proteins, ratio of shared to the the unique number of peptides between two bands\n",
    "df_shareR_all = pd.DataFrame(index=groups, columns=groups)\n",
    "df_shareR_cellular = pd.DataFrame(index=groups, columns=groups)\n",
    "df_shareR_secreted = pd.DataFrame(index=groups, columns=groups)\n",
    "for k1, k2 in itertools.product(df_group.index,repeat=2):\n",
    "    df_shareR_all.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids'] & df_group.loc[k2, 'protein_ids']) / len(df_group.loc[k1, 'protein_ids'] | df_group.loc[k2, 'protein_ids'])\n",
    "    df_shareR_cellular.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids_cellular'] & df_group.loc[k2, 'protein_ids_cellular']) / len(df_group.loc[k1, 'protein_ids_cellular'] | df_group.loc[k2, 'protein_ids_cellular'])\n",
    "    df_shareR_secreted.loc[k1,k2] = len(df_group.loc[k1, 'protein_ids_secreted'] & df_group.loc[k2, 'protein_ids_secreted']) / len(df_group.loc[k1, 'protein_ids_secreted'] | df_group.loc[k2, 'protein_ids_secreted'])\n",
    "\n",
    "df_shareR_all.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_shareR_all.csv', sep='\\t')\n",
    "df_shareR_cellular.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_shareR_cellular.csv', sep='\\t')\n",
    "df_shareR_secreted.to_csv(r'C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191027gel_bands_protein_shareR_secreted.csv', sep='\\t')\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot number of proteins in each band"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "\n",
    "xf <- \"C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191027gel_bands_protein_count.csv\"\n",
    "\n",
    "xdf<- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "\n",
    "xdf <- xdf %>%gather(key = 'type', value = 'count', - protein_count, -X)\n",
    "\n",
    "xdf['stage'] <- sapply(xdf$X, FUN = function(x){strsplit(x,'-')[[1]][1]})\n",
    "xdf$stage <- factor(xdf$stage, levels = c(\"Bar\",\"PupaD4\",\"PupaD13\",\"Adult\"))\n",
    "xdf['band'] <- sapply(xdf$X, FUN = function(x){as.numeric(strsplit(x,'-')[[1]][2])})\n",
    "\n",
    "ggplot(xdf) + geom_bar(stat = 'identity', color='transparent', aes(x=band,y=count, fill=type), position = position_dodge()) + theme_classic() + facet_grid(.~stage) + scale_fill_brewer(palette=\"Greens\")\n",
    "\n",
    "p <- ggplot(xdf) + geom_bar(stat = 'identity', color='transparent', aes(x=band,y=count, fill=type), position = position_dodge()) + theme_classic() + facet_grid(.~stage) + scale_fill_brewer(palette=\"Greens\")\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191027gel_bands_protein_count.pdf',width = 16,height = 8)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot percentage of cellular proteins in each band based on intensity\n",
    "cellular protein is defined as proteins with signal peptide\n",
    "get data\n",
    "```python\n",
    "import pandas as pd\n",
    "import itertools\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# 32 groups\n",
    "groups = []\n",
    "for stage in stages:\n",
    "    for band in range(1,9):\n",
    "        groups.append(f'{stage}-{band}')\n",
    "\n",
    "# get number of identified proteins for each group\n",
    "group_columns = ['protein_ids', 'protein_count', 'protein_ids_cellular','protein_count_cellular','protein_ids_secreted', 'protein_count_secreted']\n",
    "\n",
    "df_cellular = pd.DataFrame()\n",
    "for group in groups:\n",
    "    stage, band = group.split('-')\n",
    "    for rep in 'ABC':\n",
    "        col = f'Intensity {stage}{rep}{band}'\n",
    "        total_intensity = df[col].sum()\n",
    "        cellular_intensity = df[~df[\"SignalP\"]][col].sum()\n",
    "        df_cellular.loc[group,'total_intensity' + rep] = total_intensity\n",
    "        df_cellular.loc[group, 'cellular_intensity' + rep] = cellular_intensity\n",
    "        df_cellular.loc[group, 'cellular_percent' + rep] = cellular_intensity / total_intensity\n",
    "        df_cellular.loc[group,'stage'] = stage\n",
    "        df_cellular.loc[group,'band'] = band\n",
    "\n",
    "# calcualte average and SD\n",
    "columns_rep = [f'cellular_percent{rep}' for rep in 'ABC']\n",
    "df_cellular['cellular_percent'] = df_cellular[columns_rep].mean(axis=1)\n",
    "df_cellular['cellular_percent.sd'] = df_cellular[columns_rep].std(axis=1)\n",
    "\n",
    "# save the table\n",
    "df_cellular.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191031gel_bands_intensity_and_cellularPercent.csv\", sep='\\t',index=None)\n",
    "\n",
    "```\n",
    "\n",
    "plot\n",
    "\n",
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191031gel_bands_intensity_and_cellularPercent.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "\n",
    "xdf$stage <- factor(xdf$stage, levels = c(\"Bar\",\"PupaD4\",\"PupaD13\",\"Adult\"))\n",
    "\n",
    "test <- xdf[xdf$stage == 'Bar' & xdf$band == 1,]\n",
    "ggplot(test) + geom_bar(aes(x=band,y=cellular_percent),stat='identity', color='black',fill='transparent')+\n",
    "    geom_errorbar(aes(x=band,ymin=cellular_percent-cellular_percent.sd, ymax=cellular_percent+cellular_percent.sd), width=.2)\n",
    "\n",
    "ggplot(xdf) + geom_bar(aes(x=band,y=cellular_percent), width=.5,stat='identity', color='black',fill='transparent')+\n",
    "    geom_errorbar(aes(x=band,ymin=cellular_percent-cellular_percent.sd, ymax=cellular_percent+cellular_percent.sd), width=.2) + facet_grid(.~stage) + scale_fill_manual(values = col_vector) + theme_classic()\n",
    "\n",
    "p <- ggplot(xdf) + geom_bar(aes(x=band,y=cellular_percent), width=.5,stat='identity', color='black',fill='transparent')+\n",
    "    geom_errorbar(aes(x=band,ymin=cellular_percent-cellular_percent.sd, ymax=cellular_percent+cellular_percent.sd), width=.2) + facet_grid(.~stage) + scale_fill_manual(values = col_vector) + theme_classic()\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191031gel_bands_intensity_and_cellularPercent.pdf',width = 16,height = 8)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GO and KO summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GO summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "only check level 1 for 'biological_process', 'cellular_component', 'molecular_function'\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "\n",
    "# read in dataframe\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "\n",
    "from goatools.obo_parser import GODag\n",
    "obodag = GODag(r\"C:\\P\\Anaconda3\\Lib\\site-packages\\goatools-0.9.9-py3.7.egg\\go.obo\")\n",
    "\n",
    "# get base levels\n",
    "base_levels = [e for e in obodag.values() if e.level == 0] # 8 GO terms due to alternative ids.\n",
    "base_levels = list(set([e.id for e in base_levels])) # only keep ids that is used.\n",
    "\n",
    "# create a dataframe, with Protein and GO.\n",
    "df_GO = df[['Protein','GO']].copy()\n",
    "# use the IDs in go.obo. if a go term not in go.obo, do not include it\n",
    "def changeGO(GOs):\n",
    "    '''\n",
    "    use the IDs in go.obo. if a go term not in go.obo, do not include it\n",
    "    return a str, join by '|', the same s the \"GO\" column\n",
    "    '''\n",
    "    GOs_n = []\n",
    "    if type(GOs) is not str:\n",
    "        return ''\n",
    "    GOs = set(GOs.split('|')) # remove duplicates\n",
    "    for go in GOs:\n",
    "        go_new = obodag.query_term(go)\n",
    "        if go_new:\n",
    "            GOs_n.append(go_new.id)\n",
    "    GOs_n = set(GOs_n) # remove duplicates\n",
    "    return '|'.join(GOs_n)\n",
    "\n",
    "df_GO['GO_n'] = df_GO['GO'].apply(changeGO)\n",
    "\n",
    "\n",
    "a = obodag.paths_to_top('GO:0019318')\n",
    "\n",
    "# add three columns 'biological_process', 'cellular_component', 'molecular_function' to indicate if there is a GO annotation for the gene\n",
    "keys = ['biological_process', 'cellular_component', 'molecular_function']\n",
    "def checkKey(GOs, key):\n",
    "    '''\n",
    "    for GOs term,  check if the namespace is in the key for at least one of the GOs\n",
    "    '''\n",
    "    if GOs == '':\n",
    "        return False\n",
    "    GOs = GOs.split('|')\n",
    "    for GO in GOs:\n",
    "        if obodag[GO].namespace == key:\n",
    "            return True\n",
    "    return False\n",
    "for key in keys:\n",
    "    df_GO[key] = df_GO['GO_n'].apply(lambda x:checkKey(x,key))\n",
    "\n",
    "df_GO['withGO'] = df_GO[keys].sum(axis=1) > 0\n",
    "\n",
    "for key in keys + ['withGO']:\n",
    "    print(df_GO[key].sum(),key)\n",
    "\n",
    "level1s = [e for e in obodag.values() if e.level == 2] # 8 GO terms due to alternative ids.\n",
    "print(len(set([e.id for e in level1s]))) # only keep ids that is used.\n",
    "# count number of proteins for level 1 for 'biological_process', 'cellular_component', 'molecular_function'\n",
    "df_level1 = pd.DataFrame()\n",
    "level1s = []\n",
    "for go in base_levels:\n",
    "    for i in obodag[go].children:\n",
    "        level1s.append(i.id)\n",
    "df_level1['level1_id'] = level1s\n",
    "df_level1['GO_name'] = df_level1['level1_id'].apply(lambda x:obodag[x].name)\n",
    "df_level1['GO_type'] = df_level1['level1_id'].apply(lambda x:obodag[x].namespace)\n",
    "\n",
    "def checkInGO(GOs, goCheck):\n",
    "    '''\n",
    "    check if any one in GOs is a subfamily of goCheck\n",
    "    '''\n",
    "    if GOs == '':\n",
    "        return False\n",
    "    GOs = GOs.split('|')\n",
    "    for GO in GOs:\n",
    "        go_path = obodag.paths_to_top(GO)\n",
    "        for i in go_path:\n",
    "            for ii in i:\n",
    "                if ii.id == goCheck:\n",
    "                    return True\n",
    "    return False\n",
    "\n",
    "df_level1['protein_count'] = df_level1['level1_id'].apply(lambda key:df_GO['GO_n'].apply(lambda x:checkInGO(x,key)).sum())\n",
    "# drop columns with protein_count == 0\n",
    "df_level1 = df_level1[df_level1['protein_count'] > 0]\n",
    "df_level1.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191103ProteinsTopGO.csv\", sep='\\t', index=None)\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "some summary\n",
    "```\n",
    "871 biological_process\n",
    "432 cellular_component\n",
    "1222 molecular_function\n",
    "1349 withGO\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot GO\n",
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191103ProteinsTopGO.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "\n",
    "# remove those less than 10 proteins\n",
    "xdf <- xdf[xdf$protein_count >= 10,]\n",
    "# sort by GO_type, and protein_count\n",
    "xdf <- xdf[order(xdf$GO_type, -xdf$protein_count),]\n",
    "# to keep order, set GO_names as levels\n",
    "xdf$GO_name <- factor(xdf$GO_name, levels = xdf$GO_name)\n",
    "\n",
    "# plot\n",
    "mytheme <-theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = \"black\", size = 0.5), panel.background = element_blank(), axis.ticks = element_line(colour = \"black\", size = 0.5), axis.text = element_text(color = \"black\")) #+ them(axis.text.x = element_text(angle=45, hjust = 1))\n",
    "\n",
    "ggplot(xdf, aes(x=GO_name, y=protein_count, fill=GO_type)) + geom_bar(stat = 'identity') + mytheme + geom_text(aes(label=protein_count,),hjust=-0.1) + coord_flip() #+ theme(legend.position = \"none\")\n",
    "\n",
    "p <- ggplot(xdf, aes(x=GO_name, y=protein_count, fill=GO_type)) + geom_bar(stat = 'identity') + mytheme + geom_text(aes(label=protein_count,),hjust=-0.1) + coord_flip() #+ theme(legend.position = \"none\")\n",
    "\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191103ProteinsTopGO.pdf', width = 6, height = 4)\n",
    "p\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### high expressed proteins\n",
    "high expressed proteins are defined as proteins that is Top 5 in any gel group (stage-slice) and > 1% of total intensity in that gel group. (note: filename might be Top6, but it is Top5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### top 5 proteins in each band based on intensity\n",
    "protein with intensity > 1% of the band. calculate percentage in each replicate, then calculate average.\n",
    "\n",
    "if more than 5 protein > 1%, keep Top 5 and proteins that more than 5% in any band.\n",
    "\n",
    "```python\n",
    "import pandas as pd\n",
    "import itertools\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "outfile = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.added.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# 32 groups\n",
    "groups = []\n",
    "for stage in stages:\n",
    "    for band in range(1,9):\n",
    "        groups.append(f'{stage}-{band}')\n",
    "\n",
    "# get number of identified proteins for each group\n",
    "group_columns = ['protein_ids', 'protein_count', 'protein_ids_cellular','protein_count_cellular','protein_ids_secreted', 'protein_count_secreted']\n",
    "\n",
    "# for each group, get a dataframe of proteins with intensity > 1%. for each dataframe, keep columns \"Protein\", \"description\", \"SignalP\", \"percentage\". if there is more than 5 proteins, keep only five\n",
    "df_percent = pd.DataFrame(index=df['Protein'])\n",
    "for group in groups:\n",
    "    stage, band = group.split('-')\n",
    "    columns_rep = [f'Intensity {stage}{rep}{band}' for rep in 'ABC']\n",
    "    columns = columns_rep + [\"Protein\", \"description\", \"SignalP\", \"MolecularWeight\"]\n",
    "    tdf = df[columns].copy()\n",
    "    for col in columns_rep:\n",
    "        total_iBAQ = tdf[col].sum()\n",
    "        tdf[col+' ratio'] = tdf[col] / total_iBAQ\n",
    "    tdf['Intensity ratio'] = tdf[[col + ' ratio' for col in columns_rep]].mean(axis=1)\n",
    "    df_percent[group] = list(tdf['Intensity ratio'])\n",
    "\n",
    "df_percent['max'] = df_percent.max(axis=1)\n",
    "for i in range(10):\n",
    "    print(i, (df_percent['max']>(i/100)).sum())\n",
    "\n",
    "#always check proteins with max >= 0.5, 23 proteins\n",
    "unique_proteins = set(df_percent[df_percent['max']>0.05].index)\n",
    "print('unique_proteins to check', len(unique_proteins))\n",
    "#keep based on unique_proteins. for each band, keep those greater than 0.01\n",
    "dc_group = {}\n",
    "for group in groups:\n",
    "    stage, band = group.split('-')\n",
    "    columns_rep = [f'Intensity {stage}{rep}{band}' for rep in 'ABC']\n",
    "    columns = columns_rep + [\"Protein\", \"description\", \"SignalP\", \"MolecularWeight\"]\n",
    "    tdf = df[columns].copy()\n",
    "    for col in columns_rep:\n",
    "        total_iBAQ = tdf[col].sum()\n",
    "        tdf[col+' ratio'] = tdf[col] / total_iBAQ\n",
    "    tdf[f'Intensity ratio'] = tdf[[col + ' ratio' for col in columns_rep]].mean(axis=1)\n",
    "    tdf['group'] = group\n",
    "    column_keep = [f'Intensity ratio'] + [\"Protein\", \"description\", \"SignalP\", \"MolecularWeight\",'group']\n",
    "    tdf = tdf[column_keep]\n",
    "    tdf = tdf.sort_values(by = 'Intensity ratio', ascending=False)\n",
    "    tdf = tdf[tdf['Intensity ratio'] > 0.01]\n",
    "    if tdf.shape[0] >= 5:\n",
    "        tdf1 = tdf.iloc[:5]\n",
    "        tdf = tdf.iloc[5:]\n",
    "        tdf = tdf[tdf['Protein'].apply(lambda x:x in unique_proteins)]\n",
    "        tdf = pd.concat([tdf1,tdf])\n",
    "    tdf.loc['other','Protein'] = 'other'\n",
    "    tdf.loc['other', 'Intensity ratio'] = 1 - tdf['Intensity ratio'].sum()\n",
    "    tdf['stage'] = stage\n",
    "    tdf['band'] = int(band)\n",
    "    dc_group[group] = tdf\n",
    "\n",
    "df_groupTop = pd.concat(dc_group.values(),sort=False)\n",
    "print(df_groupTop.shape)\n",
    "print('unique_proteins to check', len(df_groupTop['Protein'].unique()))\n",
    "df_groupTop.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191029gel_bands_Top6proteinIntensity.csv\", sep='\\t', index=None)\n",
    "\n",
    "```\n",
    "\n",
    "plot with R\n",
    "\n",
    "```\n",
    "\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "\n",
    "#colors\n",
    "n <- 31\n",
    "palette <- distinctColorPalette(n)\n",
    "col_vector = c(palette, \"#FFFFFF\")\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191029gel_bands_Top6proteinIntensity.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "\n",
    "x.proteins <- xdf %>% group_by(Protein) %>% summarise(Intensity.sum = sum(Intensity.ratio)) %>% arrange(-Intensity.sum)\n",
    "x.proteins.id <- x.proteins$Protein\n",
    "x.proteins.id <- c(x.proteins.id[x.proteins.id != 'other'],'other')\n",
    "xdf$Protein <- factor(xdf$Protein, levels = x.proteins.id)\n",
    "xdf$stage <- factor(xdf$stage, levels = c(\"Bar\",\"PupaD4\",\"PupaD13\",\"Adult\"))\n",
    "\n",
    "test <- xdf[xdf$stage == 'Bar' & xdf$band == 1,]\n",
    "ggplot(test) + geom_bar(aes(x=band,y=Intensity.ratio, fill=Protein),stat='identity')\n",
    "\n",
    "ggplot(xdf) + geom_bar(aes(x=band,y=Intensity.ratio, fill=Protein),stat='identity', color='black') + facet_grid(.~stage) + scale_fill_manual(values = col_vector) + theme_classic()\n",
    "\n",
    "p <- ggplot(xdf) + geom_bar(aes(x=band,y=Intensity.ratio, fill=Protein),stat='identity', color='black') + facet_grid(.~stage) + scale_fill_manual(values = col_vector) + theme_classic()\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191029gel_bands_Top6proteinIntensity.pdf',width = 16,height = 8)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### annotation of the high-expressed proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "file_Top = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191029gel_bands_Top6proteinIntensity.csv\"\n",
    "\n",
    "# read in dataframe\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "df_top = pd.read_csv(file_Top, sep='\\t')\n",
    "\n",
    "# get unique IDs and sort by total ratio to match with orders in the figure\n",
    "df_pr = df_top.groupby('Protein')[['Intensity ratio']].sum()\n",
    "df_pr = df_pr.sort_values('Intensity ratio', ascending=False)\n",
    "df_pr = df_pr.loc[[e for e in df_pr.index if e != 'other']] # remove 'other'\n",
    "\n",
    "# add annotation\n",
    "## add annotation, this is by blast check one by one by Xiaolong\n",
    "notes = '''apolipophorins\n",
    "apolipophorin-3\n",
    "arylphorin subunit beta\n",
    "arylphorin subunit alpha \n",
    "serpin-1K\n",
    "transferrin\n",
    "insecticyanin-B\n",
    "basic juvenile hormone-suppressible protein 2\n",
    "hemolin\n",
    "27 kDa hemolymph protein-like\n",
    "fungal protease inhibitor-1-like\n",
    "cysteine proteinase inhibitor precursor\n",
    "chitinase-like protein EN03\n",
    "ommochrome-binding protein\n",
    "vitellogenin-like\n",
    "basic juvenile hormone-suppressible protein 1\n",
    "ejaculatory bulb-specific protein 3-like\n",
    "inter-alpha-trypsin inhibitor heavy chain H4-like\n",
    "phenoloxidase subunit 1\n",
    "filaggrin-2-like\n",
    "phenoloxidase subunit 2\n",
    "microvitellogenin\n",
    "apolipophorins\n",
    "protein bangles and beads-like\n",
    "plexin-A2-like\n",
    "aldo-keto reductase AKR2E4-like\n",
    "uncharacterized protein\n",
    "apolipophorins\n",
    "diapause bioclock protein\n",
    "hemocytin\n",
    "MD-2-related lipid-recognition protein-like'''.split('\\n')\n",
    "df_pr['annotation'] = notes\n",
    "## add signalP\n",
    "dc_sig = dict(zip(df['Protein'], df['SignalP']))\n",
    "df_pr['SignalP'] = [dc_sig[e] for e in df_pr.index]\n",
    "## add protein sequence\n",
    "dc_pep = dict(zip(df['Protein'], df['seq']))\n",
    "df_pr['seq'] = [dc_pep[e] for e in df_pr.index]\n",
    "## add MolecularWeight\n",
    "dc = dict(zip(df['Protein'], df['MolecularWeight']))\n",
    "df_pr['MolecularWeight'] = [dc[e] for e in df_pr.index]\n",
    "## add immuneRelated\n",
    "dc = dict(zip(df['Protein'], df['immuneRelated']))\n",
    "df_pr['immuneRelated'] = [dc[e] for e in df_pr.index]\n",
    "\n",
    "# save df_pr\n",
    "df_pr.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191103gel_bands_Top31Proteins.csv\", sep='\\t')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Total intensity for each slice. \n",
    "4 stages, 3 replicates, 8 bands, totally 96 slices. Calculate total intensity, LFQ, iBAQ for each slice. \n",
    "\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\20191016MsDev_all.xlsx\"\n",
    "\n",
    "df = pd.read_excel(filename, index_col=0)\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "\n",
    "df_summary = pd.DataFrame()\n",
    "for keyword in ['Intensity', 'iBAQ', 'LFQ intensity']:#Intensity PupaD13A1\n",
    "    for stage in stages:\n",
    "        for band in range(1,9):\n",
    "            for rep in 'ABC':\n",
    "                df_summary.loc[band,f'{keyword}_{stage}_{rep}'] = df[f'{keyword} {stage}{rep}{band}'].sum()\n",
    "\n",
    "df_summary.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191104gelSliceTotalIntensityLFQiBAQ.csv\", sep='\\t')\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### identify proteins whose expression have significant change in 4 stages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### anova test\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191104Protein_iBAQ_PPM.csv\"\n",
    "\n",
    "df_iBAQ = pd.read_csv(filename, sep='\\t', index_col=0)\n",
    "\n",
    "# stages\n",
    "stages = ['Bar', 'PupaD4', 'PupaD13', 'Adult']\n",
    "# 32 groups\n",
    "lanes = []\n",
    "for stage in stages:\n",
    "    for rep in 'ABC':\n",
    "        lanes.append(f'{stage}-{rep}')\n",
    "\n",
    "# get average PPM for each stage\n",
    "for stage in stages:\n",
    "    columns = [f'PPM {stage}-{rep}' for rep in 'ABC']\n",
    "    df_iBAQ[f'PPM {stage}-mean'] = df_iBAQ[columns].mean(axis=1)\n",
    "\n",
    "# add column \"PPM max\"\n",
    "df_iBAQ['PPM max'] = df_iBAQ[[f'PPM {stage}-mean' for stage in stages]].max(axis=1)\n",
    "\n",
    "for n in (1,10,100,1000):\n",
    "    print('number of proteins with PPM max >=', n, ':', (df_iBAQ['PPM max'] >= n).sum())\n",
    "\n",
    "# save proteins with PPM max >= 1000, to plot a heatmap showing the replicates\n",
    "df_keep = df_iBAQ[df_iBAQ['PPM max'] >= 1000]\n",
    "df_keep.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191104Protein_PPMmax_greaterThan1000.csv\", sep='\\t')\n",
    "\n",
    "# perform one-way anova for each gene, get the p-value of no change in all four stages\n",
    "from scipy import stats\n",
    "\n",
    "def proteinAnova(r):\n",
    "    '''\n",
    "    r is a row in df_iBAQ, return p-value of anova test\n",
    "    '''\n",
    "    ls = []\n",
    "    for stage in stages:\n",
    "        columns = [f'PPM {stage}-{rep}' for rep in 'ABC']\n",
    "        ls.append(list(r[columns]))\n",
    "    return stats.f_oneway(*ls)[1]\n",
    "\n",
    "df_iBAQ['p_anova'] = df_iBAQ.apply(proteinAnova, axis=1)\n",
    "\n",
    "# save df_iBAQ\n",
    "df_iBAQ.to_csv(r\"C:\\Users\\ATPs\\OneDrive\\Lab\\works\\2017DmSPSPH\\20180166ManducaMassSpectDevelopMembrane\\developHemolymph\\Tables\\20191104Protein_iBAQ_PPM_addAnova.csv\", sep='\\t')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot PPM max > 1000\n",
    "\n",
    "take a look, plot both z-score and log10 of PPM, it seems log10 is better, as the variance can be large."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "library(pheatmap)\n",
    "library(som)\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191104Protein_PPMmax_greaterThan1000.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "xmx <- as.matrix(xdf[,15:26])\n",
    "rownames(xmx) <- xdf$Protein\n",
    "# t transform\n",
    "xmx.norm <- normalize(xmx)\n",
    "xmx.log <- log10(xmx + 1)\n",
    "\n",
    "pheatmap(xmx.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T, main='SNVsCorrelation')\n",
    "pheatmap(xmx.log, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T, main='SNVsCorrelation')\n",
    "\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191104Protein_PPMmax_greaterThan1000.pdf',width = 8,height = 16)\n",
    "pheatmap(xmx.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T, main='SNVsCorrelation')\n",
    "pheatmap(xmx.log, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T, main='SNVsCorrelation')\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot heatmap to take a look. do z-transform\n",
    "use average PPM\n",
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "library(pheatmap)\n",
    "library(som)\n",
    "library(tidyr)\n",
    "library(readxl)\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191104Protein_iBAQ_PPM_addAnova.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "xf.all <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\20191016MsDev_all.xlsx'\n",
    "xf.Top31 <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191103gel_bands_Top31Proteins.csv' # top 31 proteins in gel band\n",
    "xdf.all <- read_xlsx(xf.all)\n",
    "xdf.Top31 <- read.csv(xf.Top31, sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "xv.Top31 <- xdf.Top31$Protein\n",
    "# plot abundant proteins, PPM max > 1000 or protein in xf.Top31, to include Msex2.11115-RD. otherwise, 100 top proteins\n",
    "xmx <- xdf[xdf$PPM.max > 1000 | xdf$Protein %in% xv.Top31,c('PPM.Bar.mean', 'PPM.PupaD4.mean','PPM.PupaD13.mean', 'PPM.Adult.mean')]\n",
    "\n",
    "rownames(xmx) <-  xdf[xdf$PPM.max > 1000 | xdf$Protein %in% xv.Top31,]$Protein\n",
    "# t transform\n",
    "xmx.norm <- normalize(xmx)\n",
    "xmx.norm <- as.data.frame(xmx.norm)\n",
    "colnames(xmx.norm) <- c('Bar', 'P4','P13', 'Adult')\n",
    "xmx.norm <- drop_na(xmx.norm)\n",
    "\n",
    "# annotation for signalP\n",
    "xdf.sig <- as.data.frame(xdf.all[,c('SignalP')])\n",
    "rownames(xdf.sig) <- as.array(xdf.all$Protein)\n",
    "xdf.sig$SignalP <- as.numeric(xdf.sig$SignalP)\n",
    "# annotation for p_anova\n",
    "xv.annova <- xdf$p_anova\n",
    "names(xv.annova) <- xdf$Protein\n",
    "xdf.anno <- xdf.sig\n",
    "xdf.anno$anova <- as.numeric(xv.annova[row.names(xdf.anno)] < 0.05)\n",
    "xdf.anno$log10AvePPM <- log10(rowSums(xdf[,c('PPM.Bar.mean', 'PPM.PupaD4.mean','PPM.PupaD13.mean', 'PPM.Adult.mean')])/4)\n",
    "\n",
    "pheatmap(xmx.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T, main='', annotation_row = xdf.sig)\n",
    "\n",
    "#plot for all\n",
    "xmx.all <- xdf[,c('PPM.Bar.mean', 'PPM.PupaD4.mean','PPM.PupaD13.mean', 'PPM.Adult.mean')]\n",
    "rownames(xmx.all) <-  xdf$Protein\n",
    "# split to with signalP and without signalP\n",
    "xmx.all.sig <- xmx.all[xdf.sig[rownames(xmx.all),]==1,]\n",
    "xmx.all.nosig <- xmx.all[xdf.sig[rownames(xmx.all),] != 1,]\n",
    "\n",
    "# t transform\n",
    "xmx.all.norm <- normalize(xmx.all)\n",
    "xmx.all.norm <- as.data.frame(xmx.all.norm)\n",
    "colnames(xmx.all.norm) <- c('Bar', 'P4','P13', 'Adult')\n",
    "xmx.all.norm <- drop_na(xmx.all.norm)\n",
    "\n",
    "pheatmap(xmx.all.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T, main='', annotation_row = xdf.sig)\n",
    "\n",
    "# t transform\n",
    "xmx.all.sig.norm <- normalize(xmx.all.sig)\n",
    "xmx.all.sig.norm <- as.data.frame(xmx.all.sig.norm)\n",
    "colnames(xmx.all.sig.norm) <- c('Bar', 'P4','P13', 'Adult')\n",
    "xmx.all.sig.norm <- drop_na(xmx.all.sig.norm)\n",
    "\n",
    "pheatmap(xmx.all.sig.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T, main='')\n",
    "\n",
    "# t transform\n",
    "xmx.all.nosig.norm <- normalize(xmx.all.nosig)\n",
    "xmx.all.nosig.norm <- as.data.frame(xmx.all.nosig.norm)\n",
    "colnames(xmx.all.nosig.norm) <- c('Bar', 'P4','P13', 'Adult')\n",
    "xmx.all.nosig.norm <- drop_na(xmx.all.nosig.norm)\n",
    "\n",
    "res <- pheatmap(xmx.all.nosig.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T, main='')\n",
    "\n",
    "dataN = xmx.all.nosig.norm\n",
    "out <- pheatmap(dataN, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T)\n",
    "res <- dataN[c(out$tree_row[[\"order\"]]),]\n",
    "\n",
    "\n",
    "# save figures based on if with signalP or not, Top 100 high expressed ones and save the table\n",
    "dev.off()\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191105Protein_PPM_different expression.pdf',width = 8,height = 16)\n",
    "dataN = xmx.all.sig.norm\n",
    "out <- pheatmap(dataN, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T,main='withSignal',annotation_row = xdf.anno)\n",
    "res <- dataN[c(out$tree_row[[\"order\"]]),]\n",
    "write.table(res,file = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191105Protein_PPM_different expression.withSignalP.txt',sep='\\t')\n",
    "\n",
    "dataN = xmx.all.nosig.norm\n",
    "out <- pheatmap(dataN, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T,main='withoutSignal',annotation_row = xdf.anno)\n",
    "res <- dataN[c(out$tree_row[[\"order\"]]),]\n",
    "write.table(res,file = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191105Protein_PPM_different expression.withoutSignalP.txt',sep='\\t')\n",
    "\n",
    "dataN = xmx.norm\n",
    "#color_range = c(\"#1B9E77\",'white','#D95F02')\n",
    "#ann_colors = list(SignalP=color_range, anova = color_range, log10totalPPM=color_range)\n",
    "out <- pheatmap(dataN, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T,main='Top',annotation_row = xdf.anno)\n",
    "res <- dataN[c(out$tree_row[[\"order\"]]),]\n",
    "write.table(res,file = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191105Protein_PPM_different expression.Top101.txt',sep='\\t')\n",
    "\n",
    "dev.off()\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot immune-related genes\n",
    "genes were defined by Jiang in sheet3 of 20191016MsDev_all.xlsx\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(RColorBrewer)\n",
    "library(randomcoloR)\n",
    "library(pheatmap)\n",
    "library(som)\n",
    "library(tidyr)\n",
    "library(readxl)\n",
    "library(xlsx)\n",
    "\n",
    "xf <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191104Protein_iBAQ_PPM_addAnova.csv'\n",
    "xdf <- read.csv(xf,sep='\\t',stringsAsFactors = F,header=TRUE)\n",
    "xf.all <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\20191016MsDev_all.xlsx'\n",
    "xdf.all <- read_xlsx(xf.all, sheet = \"Sheet1\")\n",
    "xdf.immune <- read_xlsx(xf.all, sheet = \"Sheet3\")\n",
    "immune.proteins <- as.vector(xdf.immune$Protein)\n",
    "immune.annotated <- as.vector(xdf.immune$Annotated)\n",
    "names(immune.annotated) <- immune.proteins\n",
    "# plot proteins in xdf.immune\n",
    "xmx <- xdf[ xdf$Protein %in% immune.proteins,c('PPM.Bar.mean', 'PPM.PupaD4.mean','PPM.PupaD13.mean', 'PPM.Adult.mean')]\n",
    "names.xmx <- xdf[xdf$Protein %in% immune.proteins,]$Protein\n",
    "rownames(xmx) <-  names.xmx\n",
    "# t transform\n",
    "xmx.norm <- normalize(xmx)\n",
    "xmx.norm <- as.data.frame(xmx.norm)\n",
    "colnames(xmx.norm) <- c('Bar', 'P4','P13', 'Adult')\n",
    "xmx.norm <- drop_na(xmx.norm)\n",
    "\n",
    "# annotation for signalP\n",
    "xdf.sig <- as.data.frame(xdf.all[,c('SignalP')])\n",
    "rownames(xdf.sig) <- as.array(xdf.all$Protein)\n",
    "xdf.sig$SignalP <- as.numeric(xdf.sig$SignalP)\n",
    "# annotation for p_anova\n",
    "xv.annova <- xdf$p_anova\n",
    "names(xv.annova) <- xdf$Protein\n",
    "xdf.anno <- xdf.sig\n",
    "xdf.anno$anova <- as.numeric(xv.annova[row.names(xdf.anno)] < 0.05)\n",
    "xdf.anno$log10AvePPM <- log10(rowSums(xdf[,c('PPM.Bar.mean', 'PPM.PupaD4.mean','PPM.PupaD13.mean', 'PPM.Adult.mean')])/4)\n",
    "\n",
    "pheatmap(xmx.norm, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = F, show_colnames = T, main='', annotation_row = xdf.anno)\n",
    "\n",
    "# save figures based on if with signalP or not, Top 100 high expressed ones and save the table\n",
    "dev.off()\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191130ImmuneRelatedProtein_PPM_differentExpression.pdf',width = 8,height = 64)\n",
    "\n",
    "dataN = xmx.norm\n",
    "#color_range = c(\"#1B9E77\",'white','#D95F02')\n",
    "#ann_colors = list(SignalP=color_range, anova = color_range, log10totalPPM=color_range)\n",
    "out <- pheatmap(dataN, border_color = NA, cluster_row = T, cluster_col = F,  show_rownames = T, show_colnames = T, main='Top',annotation_row = xdf.anno)\n",
    "res <- dataN[c(out$tree_row[[\"order\"]]),]\n",
    "res$Annotated <- immune.annotated[row.names(res)]\n",
    "res <- merge(res, xdf.anno, by=0,sort=F)\n",
    "res <- merge(res, xdf, by.x = \"Row.names\", by.y = \"Protein\",sort=F)\n",
    "write.xlsx(res,file = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191130ImmuneRelatedProtein_PPM_differentExpression.xlsx')\n",
    "\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get expression TPMs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### extract expression TPMs for all identified genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "file_proteins = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20191104Protein_iBAQ_PPM_addAnova.csv'\n",
    "\n",
    "df_proteins = pd.read_csv(file_proteins, sep='\\t', index_col=0)\n",
    "# get OGS2 protein IDs\n",
    "proteins_OGS2 = [e for e in df_proteins['Protein'] if e.startswith('Msex2')]\n",
    "# get MCOT protein IDs\n",
    "proteins_MCOT = [e for e in df_proteins['Protein'] if e.startswith('MCOT')]\n",
    "\n",
    "# read in TPM values based on OGS2\n",
    "file_OGS2 = r\"E:\\Lab\\works\\2013ManducaUtilities\\Transcriptome\\RSEM\\OGS2_20140430\\20140430RSEM_OGS2TPMave.xlsx\"\n",
    "df_OGS2 = pd.read_excel(file_OGS2, sheet_name='TPM_ave')\n",
    "\n",
    "# read in FPKM values based on MCOT\n",
    "file_MCOT = r\"E:\\Lab\\works\\2013ManducaUtilities\\Transcriptome\\RSEM\\MCOT\\20151005MCOTRSEM_FPKM.csv\"\n",
    "df_MCOT = pd.read_csv(file_MCOT)\n",
    "# get MCOT TPM values\n",
    "df_MCOT2 = df_MCOT.copy()\n",
    "df_MCOT2['transcript_id'] = df_MCOT['transcript_id'].apply(lambda x:x.split('_')[0])\n",
    "column_libraries = df_MCOT.columns[4:]\n",
    "for col in column_libraries:\n",
    "    df_MCOT2[col] = df_MCOT[col] / df_MCOT[col].sum() * 1e6\n",
    "\n",
    "# extract info\n",
    "df_1 = df_OGS2[df_OGS2['transcript_id'].isin(proteins_OGS2)]\n",
    "df_2 = df_MCOT2[df_MCOT2['transcript_id'].isin(proteins_MCOT)]\n",
    "df_keep = pd.concat([df_1,df_2], ignore_index = True,sort=False)\n",
    "\n",
    "# save the result\n",
    "file_out = 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\works\\\\2017DmSPSPH\\\\20180166ManducaMassSpectDevelopMembrane\\\\developHemolymph\\\\Tables\\\\20200127Protein1824_TPMs.csv'\n",
    "df_keep.to_csv(file_out, sep='\\t',index=None)\n",
    "```\n",
    "\n",
    "expression data from all available libraries stored in file `20200127Protein1824_TPMs.csv`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "201px",
    "width": "160px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "844px",
    "left": "0px",
    "right": "auto",
    "top": "110px",
    "width": "372px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
